<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lieven Clement" />


<title>1. Introduction to High Dimensional Data Analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HDDA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="intro.html">1. Introduction</a>
    </li>
    <li>
      <a href="svd.html">2. Singular Value Decomposition</a>
    </li>
    <li>
      <a href="svdGeometricInterpretation.html">2.3. Geometric Interpretation SVD</a>
    </li>
    <li>
      <a href="MDS_linkGramDistanceMatrix.html">2.7. Link MDS and Gram Distance Matrix</a>
    </li>
    <li>
      <a href="prediction.html">3. Prediction with High Dimensional Predictors</a>
    </li>
    <li>
      <a href="sparseSvd.html">4. Sparse Singular Value Decomposition</a>
    </li>
    <li>
      <a href="lda.html">5. Linear Discriminant Analysis</a>
    </li>
    <li>
      <a href="lsi.html">6. Large Scale Inference</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Labs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Coming soon!</li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/HDDA21">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">1. Introduction to High Dimensional Data Analysis</h1>
<h4 class="author">Lieven Clement</h4>
<h4 class="date">statOmics, Ghent University (<a href="https://statomics.github.io" class="uri">https://statomics.github.io</a>)</h4>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<ul>
<li>We live in a big data era</li>
<li>Massive Datasets on our location, surfing behavior, consumer data, social media, …</li>
<li>Life Sciences: advent of high throughput technologies has enabled us to measure brain activity (brain images) as well as the expression of thousands of genes, proteins, … for each subject or even single cell.</li>
<li>Industry: process control with many sensors that follow process in real time…</li>
<li>Data drive journalism</li>
<li>…</li>
</ul>
<p>Challenge: We have to learn from high dimensional data!</p>
<hr />
<div id="what-are-high-dimensional-data" class="section level2">
<h2><span class="header-section-number">1.1</span> What are high dimensional data?</h2>
<ul>
<li><p>We typically observe multiple variables/features (p) for each subject/experimental unit <span class="math inline">\(i=1,\ldots,n\)</span> i.e. <span class="math display">\[ \mathbf{x}_i^T=[x_{i1},\ldots,x_{ip}]\]</span></p></li>
<li><p>Multivariate statistics have a long history, but were designed for the <span class="math inline">\(n&gt;&gt;&gt;p\)</span> case,</p></li>
<li><p>Nowadays many high throughput technologies generate multivariate data with many variables (large <span class="math inline">\(p\)</span>) as compared to the number of independent replicates or samples (sample size <span class="math inline">\(n\)</span>), resulting in <strong>{high-dimensional data}</strong>, which is characterised by <span class="math display">\[
   p &gt;&gt;&gt; n.
\]</span></p></li>
<li><p>New statistical methods for dealing with the <span class="math inline">\(p &gt;&gt;&gt; n\)</span> case have been developed in the last 20 years. Some of them are adaptations of multivariate methods.</p></li>
</ul>
<hr />
<p>Issues with high-dimensional data:</p>
<ul>
<li><p><em>Computational problems</em>: large matrices, numerical accuracy of computers become an issue</p></li>
<li><p><em>Classical asymptotic theory does not hold</em> for <span class="math inline">\(p\rightarrow\infty\)</span> as <span class="math inline">\(n \rightarrow \infty\)</span></p></li>
<li><p><em>Model (or feature) selection</em> requires specialised methods to deal with the enormous number of possible models. (e.g. in linear regression with p potential predictors: <span class="math inline">\(2^p\)</span> possible models)</p></li>
<li><p>Models that can potentially depend on large-p predictors are vulnerable to <em>huge overfitting</em>.</p></li>
<li><p>In searching for associations between an outcome and large p potential exploratory variables, we are at risk to make <em>many false discoveries</em></p></li>
<li><p>The <em>Curse of Dimensionality</em> may cause a prediction model to become useless. (<span class="math inline">\(n\)</span> data points become sparse in large-<span class="math inline">\(p\)</span> dimensional space)</p></li>
</ul>
<hr />
</div>
</div>
<div id="important-tasks-in-high-dimensional-data-analysis" class="section level1">
<h1><span class="header-section-number">2</span> Important tasks in high dimensional data analysis?</h1>
<div id="example-kang-et-al.-2018s-droplet-based-scrna-seq-data-of-pbmcs-cells-from-8-lupus-patients-measured-before-and-after-6h-treatment-with-inf-beta-16-samples-in-total." class="section level2">
<h2><span class="header-section-number">2.1</span> Example: Kang et al. (2018)’s droplet-based scRNA-seq data of PBMCs cells from 8 lupus patients measured before and after 6h-treatment with INF-<span class="math inline">\(\beta\)</span> (16 samples in total).</h2>
<pre class="r"><code>library(tidyverse)

## Packages to load and visualize the single-cell data
library(ExperimentHub)
library(scater)</code></pre>
<pre class="r"><code>## Load Kang single-cell data from ExperimentHub
eh &lt;- ExperimentHub()
sce &lt;- eh[[&quot;EH2259&quot;]]

sce</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 35635 29065 
metadata(0):
assays(1): counts
rownames(35635): MIR1302-10 FAM138A ... MT-ND6 MT-CYB
rowData names(2): ENSEMBL SYMBOL
colnames(29065): AAACATACAATGCC-1 AAACATACATTTCC-1 ... TTTGCATGGTTTGG-1
  TTTGCATGTCTTAC-1
colData names(5): ind stim cluster cell multiplets
reducedDimNames(1): TSNE
mainExpName: NULL
altExpNames(0):</code></pre>
<ul>
<li>Data on gene expression of 35635 genes of 29065 cells.</li>
</ul>
</div>
<div id="data-exploration-and-dimensionality-reduction" class="section level2">
<h2><span class="header-section-number">2.2</span> Data exploration and dimensionality reduction</h2>
<ul>
<li>Visualisation is a first essential step to learn from data</li>
</ul>
<pre class="r"><code>counts(sce)[18990:19000,9:20]</code></pre>
<pre><code>11 x 12 sparse Matrix of class &quot;dgCMatrix&quot;
                                                   
RP11-467L20.10   . .   . . .   .  .  .  . .   .   .
MYRF             . .   . . .   .  .  .  . .   .   .
TMEM258          . .   3 . .   .  1  1  . .   .   .
FEN1             . .   . . .   .  .  .  . .   .   .
FADS2            . .   . . .   .  .  .  . .   .   .
FADS1            . .   . . .   .  .  .  . .   .   .
FADS3            . .   . . .   .  .  .  . .   .   .
RAB3IL1          . .   . . .   .  .  .  . .   .   .
BEST1            1 .   . . .   .  1  .  . .   1   1
FTH1           163 8 271 4 5 174 38 58 31 3 663 612
AP003733.1       . .   . . .   .  .  .  . .   .   .</code></pre>
<ul>
<li>It is impossible to learn about the structure in the data by staring at the data matrix.</li>
<li>We should be able to explore the data in a low dimensional projection</li>
</ul>
<pre class="r"><code>plotReducedDim(sce, dimred=&quot;TSNE&quot;, colour_by=&quot;cell&quot;)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code>plotReducedDim(sce, dimred=&quot;TSNE&quot;, colour_by=&quot;stim&quot;)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-3-2.png" width="672" /></p>
<ul>
<li>Note, that we see huge effect of treatment. If I see this I am always on my guard!</li>
<li>We contacted the authors and learned that all control cells were sequenced in a first run and all stimulated cells were on a second sequencing run. So the large effect might be an effect of batch!</li>
</ul>
</div>
<div id="prediction" class="section level2">
<h2><span class="header-section-number">2.3</span> Prediction</h2>
<pre class="r"><code>plotReducedDim(sce, dimred=&quot;TSNE&quot;, colour_by=&quot;cell&quot;)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<ul>
<li><p>In single cell analysis it is key to identify cell types in scRNA-seq data sets before in-depth investigations of their functional and pathological roles.</p></li>
<li><p>Use models that were build based on reference data sets to predict cell types in new data sets based on their gene expression pattern</p></li>
<li><p>Problem: we have 35635 genes at our disposal to build this prediction model!</p></li>
<li><p>Other examples</p>
<ul>
<li><p>Prediction of risk on mortality is used on a daily basis in intensive care units to prioritise patient care.</p></li>
<li><p>Facebook: predict the identity of the faces of people that are on an new image that is uploaded.</p></li>
<li><p>Netflix: Suggest movies that you would like</p></li>
</ul></li>
</ul>
</div>
<div id="large-scale-hypothesis-testing" class="section level2">
<h2><span class="header-section-number">2.4</span> Large scale hypothesis testing</h2>
<pre class="r"><code>plotReducedDim(sce, dimred=&quot;TSNE&quot;, colour_by=&quot;cell&quot;, shape_by=&quot;stim&quot;)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<ul>
<li>Which genes are differentially expressed between control and stimulated treatment (assess in each cell type)</li>
<li>For which genes we see that the differential expression according to treatment is changing according to the cell type (interaction)</li>
<li>We have to model the gene expression for each gene</li>
</ul>
<p><span class="math display">\[
\left \{
\begin{array}{lcl}
y_{ig} &amp;\sim&amp; NB(\mu_{ig}, \phi_{g})\\
E[y_{ig}] &amp;=&amp; \mu\\
\log(\mu_{ig}) &amp;=&amp; \eta_{ig}\\
\eta_{ig} &amp;=&amp; \beta_{0,g} + \sum\limits_{c=1}^C \beta_{c,g} X_{ic} + \beta_{s,g} X_{is} + \sum\limits_{c=1}^C \beta_{c:s,g} X_{ic} X_{is} + \alpha_{ig}
\end{array} \right.
\]</span></p>
<p>With</p>
<ul>
<li><p>Cell type <span class="math inline">\(c=2\ldots C\)</span> and <span class="math inline">\(X_{ic}\)</span> is a dummy variable that <span class="math inline">\(X_{ic}=1\)</span> if cell <span class="math inline">\(i\)</span> is of cell type c and <span class="math inline">\(X_{ic}=0\)</span> otherwise. Note that cell type <span class="math inline">\(c=1\)</span> is the reference cell type</p></li>
<li><p>Indicator <span class="math inline">\(X_{is}\)</span> indicates if cell <span class="math inline">\(i\)</span> was stimulated <span class="math inline">\(X_{is}=1\)</span> or not <span class="math inline">\(X_{is}=0\)</span>. So the control treatment is the reference treatment.</p></li>
</ul>
<p>Suppose we want to test if the effect of the stimulus (average difference in expression between stimulated and non stimulated cells) is different in cell type c than in the reference cell type 1?</p>
<ul>
<li><p><span class="math inline">\(H_0: \beta_{c:s,g} = 0\)</span></p></li>
<li><p><span class="math inline">\(H_1: \beta_{c:s,g} \neq 0\)</span></p></li>
<li><p>We have to assess this for 35635 genes!</p></li>
<li><p>If we assess each test at the 5% level we can expect 0.05 * 35635 = 1782 false positives.</p></li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> massive multiple testing problem!</p>
<p>Note, that we cannot differentiate between batch and treatment because of the flaw in the experimental design!</p>
<p>Other examples</p>
<ul>
<li>Find regions (voxels) in the brain (on brain image) that are associated with a certain condition/treatment</li>
<li>Evaluation of trading rules</li>
</ul>
<hr />
</div>
</div>
<div id="linear-regression" class="section level1">
<h1><span class="header-section-number">3</span> Linear regression</h1>
<ul>
<li>Linear regression is a very important statistical tool to study the association between variables and to build prediction models.</li>
</ul>
<div id="scalar-form" class="section level2">
<h2><span class="header-section-number">3.1</span> Scalar form</h2>
<ul>
<li>Consider a vector of predictors <span class="math inline">\(\mathbf{x}=(x_1,\ldots,x_p)\)</span> and</li>
<li>a real-valued response <span class="math inline">\(Y\)</span></li>
<li>then the linear regression model can be written as <span class="math display">\[
Y=f(\mathbf{x}) +\epsilon=\beta_0+\sum\limits_{j=1}^p x_j\beta_j + \epsilon
\]</span> with i.i.d. <span class="math inline">\(\epsilon\sim N(0,\sigma^2)\)</span></li>
<li>We will typically work on mean centered variables: <span class="math inline">\(Y-\bar{Y}\)</span> and <span class="math inline">\(X - \bar{X}\)</span>, then the intercept is dropped from the model: <span class="math display">\[
Y=f(\mathbf{x}) +\epsilon=\sum\limits_{j=1}^p x_j\beta_j + \epsilon
\]</span></li>
</ul>
</div>
<div id="vectormatrix-form" class="section level2">
<h2><span class="header-section-number">3.2</span> Vector/Matrix form</h2>
<ul>
<li><span class="math inline">\(n\)</span> observations <span class="math inline">\((\mathbf{x}_1,y_1) \ldots (\mathbf{x}_n,y_n)\)</span> with <span class="math inline">\(\mathbf{x}_1^T=[1 x_1 \ldots x_p]\)</span></li>
<li>Regression in matrix notation <span class="math display">\[\mathbf{Y}=\mathbf{X\beta} + \mathbf{\epsilon}\]</span> with <span class="math inline">\(\mathbf{Y}=\left[\begin{array}{c}y_1\\ \vdots\\y_n\end{array}\right]\)</span>, <span class="math inline">\(\mathbf{X}=\left[\begin{array}{cccc} 1&amp;x_{11}&amp;\ldots&amp;x_{1p}\\ \vdots&amp;\vdots&amp;&amp;\vdots\\ 1&amp;x_{n1}&amp;\ldots&amp;x_{np} \end{array}\right]\)</span> or <span class="math inline">\(\mathbf{X}=\left[\begin{array}{c} \mathbf{x}_1^T\\\vdots\\\mathbf{x}_n^T\end{array}\right]\)</span>, <span class="math inline">\(\boldsymbol{\beta}=\left[\begin{array}{c}\beta_0\\ \vdots\\ \beta_p\end{array}\right]\)</span> and <span class="math inline">\(\mathbf{\epsilon}=\left[\begin{array}{c} \epsilon_1 \\ \vdots \\ \epsilon_n\end{array}\right]\)</span></li>
</ul>
<p>Note, that upon centering of <span class="math inline">\(\mathbf{X}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span> the 1 is dropped from each <span class="math inline">\(\mathbf{x}_i\)</span> and thus the column of 1 is dropped in <span class="math inline">\(\mathbf{X}\)</span></p>
<hr />
</div>
<div id="interpretation" class="section level2">
<h2><span class="header-section-number">3.3</span> Interpretation</h2>
<p>From the linear regression we get <span class="math display">\[
  E[Y \mid \mathbf{x}] = \mathbf{x}^T\boldsymbol{\beta} .
\]</span></p>
<ul>
<li>Hence, the <span class="math inline">\(\beta\)</span> parameters relate the regressor <span class="math inline">\(\mathbf{x}\)</span> to the mean outcome.</li>
<li>If we know the covariate pattern <span class="math inline">\(\mathbf{x}\)</span> we can use the model to predict <span class="math inline">\(Y\)</span>.</li>
</ul>
<p>Upon centering we get the following for a model with a single regressor <span class="math display">\[
  E[Y \mid x] = \beta_1 x
\]</span> and <span class="math display">\[
 \beta_1 = E[Y\mid x+1] - E[Y\mid x] .
\]</span></p>
<p>Hence, the <span class="math inline">\(\beta_1\)</span> parameter has an interpretation as the mean increase of the outcome as the regressor increases with one unit. The parameter does not say very about individual outcomes. It is the residual variance <span class="math inline">\(\sigma^2\)</span> that determines how much individual outcomes vary about the mean outcome.</p>
<p>The <span class="math inline">\(\beta\)</span> parameters are used to measure association, but a <span class="math inline">\(\beta \neq 0\)</span> does not necessarily mean that the model will give good predictions.</p>
<hr />
<ul>
<li><p>In a later chapter we will discuss the problem of large scale hypothesis testing: testing many hypotheses in a single study (ten to hundred thousands of hypotheses).</p></li>
<li><p>A statistical test is constructed to control the type I error rate at the significance level. However, when many hypotheses are to be tested in a single study, the probability to find false associations is no longer controlled if p-values are compared to the significance level.</p></li>
<li><p>We will later introduce the concept of false discovery rates to overcome the problem.</p></li>
</ul>
</div>
<div id="least-squares-ls" class="section level2">
<h2><span class="header-section-number">3.4</span> Least Squares (LS)</h2>
<ul>
<li><p>Minimize the residual sum of squares <span class="math display">\[\begin{eqnarray*}
RSS(\boldsymbol{\beta})&amp;=&amp;\sum\limits_{i=1}^n e^2_i\\
&amp;=&amp;\sum\limits_{i=1}^n \left(y_i-\beta_0-\sum\limits_{j=1}^p x_{ij}\beta_j\right)^2
\end{eqnarray*}\]</span></p></li>
<li><p>or in matrix notation <span class="math display">\[\begin{eqnarray*}
RSS(\boldsymbol{\beta})&amp;=&amp;(\mathbf{Y}-\mathbf{X\beta})^T(\mathbf{Y}-\mathbf{X\beta})\\
&amp;=&amp;\Vert \mathbf{Y}-\mathbf{X\beta}\Vert^2_2
\end{eqnarray*}\]</span> with the <span class="math inline">\(L_2\)</span>-norm of a <span class="math inline">\(p\)</span>-dim. vector <span class="math inline">\(v\)</span> <span class="math inline">\(\Vert \mathbf{v} \Vert_2=\sqrt{v_1^2+\ldots+v_p^2}\)</span> <span class="math inline">\(\rightarrow\)</span> <span class="math inline">\(\hat{\boldsymbol{\beta}}=\text{argmin}_\beta \Vert \mathbf{Y}-\mathbf{X\beta}\Vert^2\)</span></p></li>
</ul>
<hr />
<div id="minimize-rss" class="section level3">
<h3><span class="header-section-number">3.4.1</span> Minimize RSS</h3>
<p><span class="math display">\[
\begin{array}{ccc}
\frac{\partial RSS}{\partial \boldsymbol{\beta}}&amp;=&amp;\mathbf{0}\\\\
\frac{(\mathbf{Y}-\mathbf{X\beta})^T(\mathbf{Y}-\mathbf{X}\boldsymbol{\beta})}{\partial \boldsymbol{\beta}}&amp;=&amp;\mathbf{0}\\\\
-2\mathbf{X}^T(\mathbf{Y}-\mathbf{X}\boldsymbol{\beta})&amp;=&amp;\mathbf{0}\\\\
\mathbf{X}^T\mathbf{X\beta}&amp;=&amp;\mathbf{X}^T\mathbf{Y}\\\\
\hat{\boldsymbol{\beta}}&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{Y}
\end{array}
\]</span></p>
<p>It can be shown that the estimator is unbiased:</p>
<p><span class="math display">\[
E[\hat{\boldsymbol{\beta}}]=\boldsymbol{\beta}
\]</span></p>
<hr />
</div>
<div id="geometrical-interpretation" class="section level3">
<h3><span class="header-section-number">3.4.2</span> Geometrical Interpretation</h3>
<p>Toy Example: fit without intercept</p>
<ul>
<li>n=3 and p=2 <span class="math display">\[
\mathbf{X}=\left[\begin{array}{cc}
2&amp;0\\
0&amp;2\\
0&amp;0
\end{array}\right]
\]</span></li>
</ul>
<pre class="r"><code>set.seed(4)
x1 &lt;- c(2, 0, 0)
x2 &lt;- c(0, 2, 0)
y &lt;- x1 * 0.5 + x2 * 0.5 + rnorm(3, 2)
fit &lt;- lm(y ~ -1 + x1 + x2)</code></pre>
<ul>
<li><span class="math inline">\(\mathbf{Y}\)</span> is now given by</li>
</ul>
<p><span class="math display">\[
\mathbf{Y}=\left[\begin{array}{c}
1 + \epsilon_1 \\
1 + \epsilon_2\\
0 + \epsilon_3
\end{array}\right]
\]</span></p>
<p>with the <span class="math inline">\(\epsilon_i\)</span>s some random noise.</p>
<ul>
<li>The actual values of <span class="math inline">\(\mathbf{Y}\)</span> are:</li>
</ul>
<pre class="r"><code>## True Y
y</code></pre>
<pre><code>[1] 3.216755 2.457507 2.891145</code></pre>
<pre class="r"><code>## Fitted Y
fit$fitted.values</code></pre>
<pre><code>       1        2        3 
3.216755 2.457507 0.000000 </code></pre>
<div id="visualise-fit" class="section level4">
<h4><span class="header-section-number">3.4.2.1</span> Visualise fit</h4>
<pre class="r"><code># predict values on regular xy grid
x1pred &lt;- seq(-1, 3, length.out = 10)
x2pred &lt;- seq(-1, 3, length.out = 10)
xy &lt;- expand.grid(x1 = x1pred,
x2 = x2pred)
ypred &lt;- matrix (nrow = 30, ncol = 30,
data = predict(fit, newdata = data.frame(xy),
interval = &quot;prediction&quot;))

library(plot3D)


# fitted points for droplines to surface
th=20
ph=5
scatter3D(x1,
  x2,
  y,
  pch = 16,
  col=&quot;darkblue&quot;,
  cex = 1,
  theta = th,
  ticktype = &quot;detailed&quot;,
  xlab = &quot;x1&quot;,
  ylab = &quot;x2&quot;,
  zlab = &quot;y&quot;,
  colvar=FALSE,
  bty = &quot;g&quot;,
  xlim=c(-1,3),
  ylim=c(-1,3),
  zlim=c(-2,4))

for (i in 1:3)
  lines3D(
    x = rep(x1[i],2),
    y = rep(x2[i],2),
    z = c(y[i],fit$fitted[i]),
    col=&quot;darkblue&quot;,
    add=TRUE,
    lty=2)

z.pred3D &lt;- outer(
  x1pred,
  x2pred,
  function(x1,x2)
  {
    fit$coef[1]*x1+fit$coef[2]*x2
  })

x.pred3D &lt;- outer(
  x1pred,
  x2pred,
  function(x,y) x)

y.pred3D &lt;- outer(
  x1pred,
  x2pred,
  function(x,y) y)

surf3D(
  x.pred3D,
  y.pred3D,
  z.pred3D,
  col=&quot;blue&quot;,
  facets=NA,
  add=TRUE)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="projection" class="section level4">
<h4><span class="header-section-number">3.4.2.2</span> Projection</h4>
<ul>
<li><p>We can also interpret the fit as the projection of the <span class="math inline">\(n\times 1\)</span> vector <span class="math inline">\(\mathbf{Y}\)</span> on the column space of the matrix <span class="math inline">\(\mathbf{X}\)</span>.</p></li>
<li><p>So each column in <span class="math inline">\(\mathbf{X}\)</span> is also an <span class="math inline">\(n\times 1\)</span> vector.</p></li>
<li><p>For the toy example n=3 and p=2. So the column space of X is a plane in the three dimensional space.</p></li>
</ul>
<p><span class="math display">\[
\hat{\mathbf{Y}} = \mathbf{X} (\mathbf{X}^T\mathbf{X})^{-1} \mathbf{X}^T \mathbf{Y}
\]</span></p>
<ol style="list-style-type: decimal">
<li>Plane spanned by column space:</li>
</ol>
<pre class="r"><code>arrows3D(
  0, 0, 0, x1[1], x1[2], x1[3],
  xlim = c(0, 5), ylim = c(0, 5), zlim = c(0, 5), bty = &quot;g&quot;,
  theta = th, col = 2, ticktype = &quot;detailed&quot;,
  xlab = &quot;observation 1&quot;, ylab = &quot;observation 2&quot;, zlab = &quot;observation 3&quot;
)
text3D(x1[1],x1[2],x1[3],labels=&quot;X1&quot;,col=2,add=TRUE)
arrows3D(0,0,0,x2[1],x2[2],x2[3],add=TRUE,col=2)
text3D(x2[1],x2[2],x2[3],labels=&quot;X2&quot;,col=2,add=TRUE)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<ol start="2" style="list-style-type: decimal">
<li>Vector of Y:</li>
</ol>
<p>Actual values of <span class="math inline">\(\mathbf{Y}\)</span>:</p>
<pre class="r"><code>y</code></pre>
<pre><code>[1] 3.216755 2.457507 2.891145</code></pre>
<p><span class="math display">\[
\mathbf{Y}=\left[\begin{array}{c}
3.2167549 \\
2.4575074 \\
2.8911446
\end{array}\right]
\]</span></p>
<pre class="r"><code>arrows3D(
  0, 0, 0, x1[1], x1[2], x1[3],
  xlim = c(0, 5), ylim = c(0, 5), zlim = c(0, 5), bty = &quot;g&quot;,
  theta = th, col = 2, ticktype = &quot;detailed&quot;,
  xlab = &quot;observation 1&quot;, ylab = &quot;observation 2&quot;, zlab = &quot;observation 3&quot;
)
text3D(x1[1],x1[2],x1[3],labels=&quot;X1&quot;,col=2,add=TRUE)
arrows3D(0,0,0,x2[1],x2[2],x2[3],add=TRUE,col=2)
text3D(x2[1],x2[2],x2[3],labels=&quot;X2&quot;,col=2,add=TRUE)
arrows3D(0,0,0,y[1],y[2],y[3],add=TRUE,col=&quot;darkblue&quot;)
text3D(y[1],y[2],y[3],labels=&quot;Y&quot;,col=&quot;darkblue&quot;,add=TRUE)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<ol start="3" style="list-style-type: decimal">
<li>Projection of Y onto column space</li>
</ol>
<p>Actual values of fitted values <span class="math inline">\(\mathbf{\hat{Y}}\)</span>:</p>
<pre class="r"><code>fitted.values(fit)</code></pre>
<pre><code>       1        2        3 
3.216755 2.457507 0.000000 </code></pre>
<p><span class="math display">\[
\mathbf{Y}=\left[\begin{array}{c}
3.2167549 \\
2.4575074 \\
0
\end{array}\right]
\]</span></p>
<pre class="r"><code>arrows3D(
  0, 0, 0, x1[1], x1[2], x1[3],
  xlim = c(0, 5), ylim = c(0, 5), zlim = c(0, 5), bty = &quot;g&quot;,
  theta = th, col = 2, ticktype = &quot;detailed&quot;,
  xlab = &quot;observation 1&quot;, ylab = &quot;observation 2&quot;, zlab = &quot;observation 3&quot;
)
text3D(x1[1],x1[2],x1[3],labels=&quot;X1&quot;,col=2,add=TRUE)
arrows3D(0,0,0,x2[1],x2[2],x2[3],add=TRUE,col=2)
text3D(x2[1],x2[2],x2[3],labels=&quot;X2&quot;,col=2,add=TRUE)
arrows3D(0,0,0,y[1],y[2],y[3],add=TRUE,col=&quot;darkblue&quot;)
text3D(y[1],y[2],y[3],labels=&quot;Y&quot;,col=&quot;darkblue&quot;,add=TRUE)
arrows3D(0,0,0,fit$fitted[1],fit$fitted[2],fit$fitted[3],add=TRUE,col=&quot;darkblue&quot;)
segments3D(y[1],y[2],y[3],fit$fitted[1],fit$fitted[2],fit$fitted[3],add=TRUE,lty=2,col=&quot;darkblue&quot;)
text3D(fit$fitted[1],fit$fitted[2],fit$fitted[3],labels=&quot;fit&quot;,col=&quot;darkblue&quot;,add=TRUE)</code></pre>
<p><img src="intro_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<ul>
<li><p>Note, that it is also clear from the equation in the derivation of the LS that the residual is orthogonal on the column space: <span class="math display">\[
 -2 \mathbf{X}^T(\mathbf{Y}-\mathbf{X}\boldsymbol{\beta}) = 0
\]</span></p></li>
<li><p>Imagine what happens when p &gt;&gt; n!!!</p></li>
</ul>
<p><span class="math inline">\(\rightarrow\)</span> curse of dimensionality!</p>
</div>
</div>
</div>
<div id="variance-estimator" class="section level2">
<h2><span class="header-section-number">3.5</span> Variance Estimator?</h2>
<p><span class="math display">\[
\begin{array}{ccl}
\hat{\boldsymbol{\Sigma}}_{\hat{\boldsymbol{\beta}}}
&amp;=&amp;\text{var}\left[(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{Y}\right]\\\\
&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\text{var}\left[\mathbf{Y}\right]\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}\\\\
&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T(\mathbf{I}\sigma^2)\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}
\\\\
&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{I}\quad\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2\\\\
%\hat{\boldmath{\Sigma}}_{\hat{\boldsymbol{\beta}}}&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T \text{var}\left[\mathbf{Y}\right](\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}\\
&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2\\\\
&amp;=&amp;(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2
\end{array}
\]</span></p>
<p>The fact that <span class="math inline">\(\hat{\boldsymbol{\beta}}\)</span> is unbiased and has a variance of <span class="math inline">\((\mathbf{X}^T\mathbf{X})^{-1}\sigma^2\)</span> will be important when assessing association!</p>
<hr />
</div>
<div id="prediction-error" class="section level2">
<h2><span class="header-section-number">3.6</span> Prediction error</h2>
<p>Least squares estimators are unbiased and consistent, but these properties are not very important for prediction models.</p>
<p>A prediction model is considered good if it can predict well outcomes.</p>
<p>The <strong>prediction error</strong> for a prediction at predictor <span class="math inline">\(\mathbf{x}\)</span> is given by</p>
<p><span class="math display">\[
   \hat{Y}(\mathbf{x}) - Y^*,
\]</span></p>
<p>where</p>
<ul>
<li><p><span class="math inline">\(\hat{Y}(\mathbf{x})=\mathbf{x}^T\hat{\boldsymbol{\beta}}\)</span> is the prediction at <span class="math inline">\(\mathbf{x}\)</span></p></li>
<li><p><span class="math inline">\(Y^*\)</span> is an outcome at predictor <span class="math inline">\(\mathbf{x}\)</span></p></li>
</ul>
<p>Since prediction is typically used to predict an outcome before it is observed, the outcome <span class="math inline">\(Y^*\)</span> is not observed yet. Hence, the prediction error cannot be computed.</p>
<hr />
<p>The problem of unobservable prediction errors is partly solved by the <strong>expected conditional test error</strong> (sometimes referred to as the mean squared error, MSE) <span class="math display">\[
  \text{Err}(\mathbf{x})=E[(\hat{Y}(\mathbf{x}) - Y^*)^2].
\]</span></p>
<p>With (suppressing the dependence on <span class="math inline">\(\mathbf{x}\)</span>) <span class="math display">\[
  \mu = E[\hat{Y}] \text{ and } \mu^*=E[Y^*]
\]</span> the error can be expressed as <span class="math display">\[\begin{eqnarray*}
  \text{Err}
    &amp;=&amp;E\left\{\left[(\hat{Y} -\mu)-(Y-\mu^*)-(\mu^*-\mu)\right]^2\right\} \\
    &amp;=&amp; E[(\hat{Y} -\mu)^2]+E[(Y-\mu^*)^2]+E[(\mu^*-\mu)^2] \\
    &amp;=&amp; \text{var}[\hat{Y}] +\text{var}[Y] +\text{bias}^2
\end{eqnarray*}\]</span></p>
<p>The term <span class="math inline">\(\text{var}[Y]\)</span> (irreducible error) does not depend on the model and may therefore be ignored when <span class="math inline">\(\text{Err}\)</span> is used for comparing prediction models.</p>
<hr />
<p>In this introductory chapter we only aim to give a rough discussion on prediction errors. Later definitions will be refined and the notation will be more accurate. Also a more detailed discussion on the bias-variance trade-off will follow. For the moment it is sufficient to vaguely know that:</p>
<ul>
<li><p>the expected conditional test error is introduced to circumvent the problem that the prediction error cannot be observed. In later chapters we will look at estimators of the expected error.</p></li>
<li><p>the expected conditional test error is in some literature also known as the <em>mean squared error</em> (MSE), but we do not adopt this terminology because MSE is also commonly used to refer to SSE divided by the residual degrees of freedom in a linear regression model.</p></li>
<li><p>The identity <span class="math inline">\(\text{Err}=\text{var}[\hat{Y}] +\text{var}[Y] +\text{bias}^2\)</span> is known as the bias-variance trade-off. It shows that a good prediction model (i.e. a model resulting in a small Err), can be obtained by a model that shows a small bias as long as this bias is compensated with a large reduction of the variance or the predictions. A more detailed discussion will follow in later chapters.</p></li>
<li><p>For prediction models with a large number of predictors we will therefore introduce penalized regression. This will induce some bias in the estimation, but will allow us to reduce the variance considerably.</p></li>
</ul>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1>Session info</h1>
<details>
<p><summary>Session info</summary></p>
<pre><code>[1] &quot;2021-10-04 13:36:56 UTC&quot;</code></pre>
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 4.1.1 (2021-08-10)
 os       macOS Catalina 10.15.7      
 system   x86_64, darwin17.0          
 ui       X11                         
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       UTC                         
 date     2021-10-04                  

─ Packages ───────────────────────────────────────────────────────────────────
 ! package                * version  date       lib
 P AnnotationDbi            1.54.1   2021-06-08 [?]
 P AnnotationHub          * 3.0.1    2021-06-20 [?]
 P assertthat               0.2.1    2019-03-21 [?]
 P backports                1.2.1    2020-12-09 [?]
 P beachmat                 2.8.1    2021-08-12 [?]
 P beeswarm                 0.4.0    2021-06-01 [?]
 P Biobase                * 2.52.0   2021-05-19 [?]
 P BiocFileCache          * 2.0.0    2021-05-19 [?]
 P BiocGenerics           * 0.38.0   2021-05-19 [?]
 P BiocManager              1.30.16  2021-06-15 [?]
 P BiocNeighbors            1.10.0   2021-05-19 [?]
 P BiocParallel             1.26.2   2021-08-22 [?]
 P BiocSingular             1.8.1    2021-06-08 [?]
 P BiocVersion              3.13.1   2021-03-09 [?]
 P Biostrings               2.60.2   2021-08-05 [?]
 P bit                      4.0.4    2020-08-04 [?]
 P bit64                    4.0.5    2020-08-30 [?]
 P bitops                   1.0-7    2021-04-24 [?]
 P blob                     1.2.2    2021-07-23 [?]
 P broom                    0.7.9    2021-07-27 [?]
 P bslib                    0.3.0    2021-09-02 [?]
 P cachem                   1.0.6    2021-08-19 [?]
 P cellranger               1.1.0    2016-07-27 [?]
 P cli                      3.0.1    2021-07-17 [?]
 P colorspace               2.0-2    2021-06-24 [?]
 P crayon                   1.4.1    2021-02-08 [?]
 P curl                     4.3.2    2021-06-23 [?]
 P DBI                      1.1.1    2021-01-15 [?]
 P dbplyr                 * 2.1.1    2021-04-06 [?]
 P DelayedArray             0.18.0   2021-05-19 [?]
 P DelayedMatrixStats       1.14.3   2021-08-26 [?]
 P digest                   0.6.28   2021-09-23 [?]
 P dplyr                  * 1.0.7    2021-06-18 [?]
 P ellipsis                 0.3.2    2021-04-29 [?]
 P evaluate                 0.14     2019-05-28 [?]
 P ExperimentHub          * 2.0.0    2021-05-19 [?]
 P fansi                    0.5.0    2021-05-25 [?]
 P farver                   2.1.0    2021-02-28 [?]
 P fastmap                  1.1.0    2021-01-25 [?]
 P filelock                 1.0.2    2018-10-05 [?]
 P forcats                * 0.5.1    2021-01-27 [?]
 P fs                       1.5.0    2020-07-31 [?]
 P generics                 0.1.0    2020-10-31 [?]
 P GenomeInfoDb           * 1.28.4   2021-09-09 [?]
 P GenomeInfoDbData         1.2.6    2021-10-04 [?]
 P GenomicRanges          * 1.44.0   2021-05-19 [?]
 P ggbeeswarm               0.6.0    2017-08-07 [?]
 P ggplot2                * 3.3.5    2021-06-25 [?]
 P glue                     1.4.2    2020-08-27 [?]
 P gridExtra                2.3      2017-09-09 [?]
 P gtable                   0.3.0    2019-03-25 [?]
 P haven                    2.4.3    2021-08-04 [?]
 P highr                    0.9      2021-04-16 [?]
 P hms                      1.1.1    2021-09-26 [?]
 P htmltools                0.5.2    2021-08-25 [?]
 P httpuv                   1.6.3    2021-09-09 [?]
 P httr                     1.4.2    2020-07-20 [?]
 P interactiveDisplayBase   1.30.0   2021-05-19 [?]
 P IRanges                * 2.26.0   2021-05-19 [?]
 P irlba                    2.3.3    2019-02-05 [?]
 P jquerylib                0.1.4    2021-04-26 [?]
 P jsonlite                 1.7.2    2020-12-09 [?]
 P KEGGREST                 1.32.0   2021-05-19 [?]
 P knitr                    1.33     2021-04-24 [?]
 P labeling                 0.4.2    2020-10-20 [?]
 P later                    1.3.0    2021-08-18 [?]
 P lattice                  0.20-45  2021-09-22 [?]
 P lifecycle                1.0.1    2021-09-24 [?]
 P lubridate                1.7.10   2021-02-26 [?]
 P magrittr                 2.0.1    2020-11-17 [?]
 P Matrix                   1.3-4    2021-06-01 [?]
 P MatrixGenerics         * 1.4.3    2021-08-26 [?]
 P matrixStats            * 0.60.1   2021-08-23 [?]
 P memoise                  2.0.0    2021-01-26 [?]
 P mime                     0.12     2021-09-28 [?]
 P misc3d                   0.9-0    2020-09-06 [?]
 P modelr                   0.1.8    2020-05-19 [?]
 P munsell                  0.5.0    2018-06-12 [?]
   muscData               * 1.6.0    2021-05-20 [1]
 P pillar                   1.6.3    2021-09-26 [?]
 P pkgconfig                2.0.3    2021-10-04 [?]
 P plot3D                 * 1.4      2021-05-22 [?]
 P png                      0.1-7    2013-12-03 [?]
 P promises                 1.2.0.1  2021-02-11 [?]
 P purrr                  * 0.3.4    2020-04-17 [?]
 P R6                       2.5.1    2021-08-19 [?]
 P rappdirs                 0.3.3    2021-01-31 [?]
 P Rcpp                     1.0.7    2021-07-07 [?]
 P RCurl                    1.98-1.4 2021-08-17 [?]
 P readr                  * 2.0.1    2021-08-10 [?]
 P readxl                   1.3.1    2019-03-13 [?]
 P renv                     0.14.0   2021-07-21 [?]
 P reprex                   2.0.1    2021-08-05 [?]
 P rlang                    0.4.11   2021-04-30 [?]
 P rmarkdown                2.10     2021-08-06 [?]
 P RSQLite                  2.2.8    2021-08-21 [?]
 P rstudioapi               0.13     2020-11-12 [?]
 P rsvd                     1.0.5    2021-04-16 [?]
 P rvest                    1.0.1    2021-07-26 [?]
 P S4Vectors              * 0.30.0   2021-05-19 [?]
 P sass                     0.4.0    2021-05-12 [?]
 P ScaledMatrix             1.0.0    2021-05-19 [?]
 P scales                   1.1.1    2020-05-11 [?]
 P scater                 * 1.20.1   2021-06-15 [?]
 P scuttle                * 1.2.1    2021-08-05 [?]
 P sessioninfo              1.1.1    2018-11-05 [?]
 P shiny                    1.7.0    2021-09-22 [?]
 P SingleCellExperiment   * 1.14.1   2021-05-21 [?]
 P sparseMatrixStats        1.4.2    2021-08-08 [?]
 P stringi                  1.7.4    2021-08-25 [?]
 P stringr                * 1.4.0    2019-02-10 [?]
 P SummarizedExperiment   * 1.22.0   2021-05-19 [?]
 P tibble                 * 3.1.5    2021-09-30 [?]
 P tidyr                  * 1.1.4    2021-09-27 [?]
 P tidyselect               1.1.1    2021-04-30 [?]
 P tidyverse              * 1.3.1    2021-04-15 [?]
 P tzdb                     0.1.2    2021-07-20 [?]
 P utf8                     1.2.2    2021-07-24 [?]
 P vctrs                    0.3.8    2021-04-29 [?]
 P vipor                    0.4.5    2017-03-22 [?]
 P viridis                  0.6.1    2021-05-11 [?]
 P viridisLite              0.4.0    2021-04-13 [?]
 P withr                    2.4.2    2021-04-18 [?]
 P xfun                     0.25     2021-08-06 [?]
 P xml2                     1.3.2    2020-04-23 [?]
 P xtable                   1.8-4    2019-04-21 [?]
 P XVector                  0.32.0   2021-05-19 [?]
 P yaml                     2.2.1    2020-02-01 [?]
 P zlibbioc                 1.38.0   2021-05-19 [?]
 source                          
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 Bioconductor                    
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 Bioconductor                    
 Bioconductor                    
 Bioconductor                    
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 standard (@1.0-7)               
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 standard (@0.3.2)               
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.1)                  
 Bioconductor                    
 CRAN (R 4.1.1)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 Github (r-lib/pkgconfig@b81ae03)
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.1)                  
 CRAN (R 4.1.1)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.1)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.1)                  
 Bioconductor                    
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 standard (@1.1.1)               
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 standard (@0.3.8)               
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.1)                  
 CRAN (R 4.1.0)                  
 CRAN (R 4.1.0)                  
 Bioconductor                    
 CRAN (R 4.1.0)                  
 Bioconductor                    

[1] /Users/runner/work/HDDA21/HDDA21/renv/library/R-4.1/x86_64-apple-darwin17.0
[2] /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/Rtmpg6sHvH/renv-system-library

 P ── Loaded and on-disk path mismatch.</code></pre>
</details>
</div>

<div id="rmd-source-code">---
title: "1. Introduction to High Dimensional Data Analysis"
author: "Lieven Clement"
date: "statOmics, Ghent University (https://statomics.github.io)"
---

# Introduction

```{r setup, include=FALSE}
knitr::opts_chunk$set(comment = NA, message = FALSE, warning = FALSE)
```

- We live in a big data era
- Massive Datasets on our location, surfing behavior, consumer data, social media, ...
- Life Sciences: advent of high throughput technologies has enabled us to measure brain activity (brain images) as well as the expression of thousands of genes, proteins, ... for each subject or even single cell.
- Industry: process control with many sensors that follow process in real time...
- Data drive journalism
- ...

Challenge: We have to learn from high dimensional data!

---

## What are high dimensional data?

- We typically observe multiple variables/features (p) for each subject/experimental unit $i=1,\ldots,n$ i.e.
\[ \mathbf{x}_i^T=[x_{i1},\ldots,x_{ip}]\]

- Multivariate statistics have a long history, but were designed for the $n>>>p$ case,

- Nowadays many high throughput technologies generate multivariate data with many variables (large $p$) as compared to the number of independent replicates or samples (sample size $n$), resulting in **{high-dimensional data}**, which is characterised by
  \[
     p >>> n.
  \]

-  New statistical methods for dealing with the $p >>> n$ case have been developed in the last 20 years.  Some of them are adaptations of multivariate methods.

---

Issues with high-dimensional data:

- *Computational problems*: large matrices, numerical accuracy of computers become an issue
- *Classical asymptotic theory does not hold* for $p\rightarrow\infty$ as $n \rightarrow \infty$
- *Model (or feature) selection* requires specialised methods to deal with the enormous number of possible models. (e.g. in linear regression with p potential predictors: $2^p$ possible models)
- Models that can potentially depend on large-p predictors are vulnerable to *huge overfitting*.
- In searching for associations between an outcome and large p potential exploratory variables, we are at risk to make *many false discoveries*

- The *Curse of Dimensionality* may cause a prediction model to become useless. ($n$ data points become sparse in large-$p$ dimensional space)

---

# Important tasks in high dimensional data analysis?

## Example:   Kang et al. (2018)’s droplet-based scRNA-seq data of PBMCs cells from 8 lupus patients measured before and after 6h-treatment with INF-$\beta$ (16 samples in total).

```{r libraries, message=FALSE, warning=FALSE}
library(tidyverse)

## Packages to load and visualize the single-cell data
library(ExperimentHub)
library(scater)
```

```{r}
## Load Kang single-cell data from ExperimentHub
eh <- ExperimentHub()
sce <- eh[["EH2259"]]

sce
```

- Data on gene expression of `r nrow(sce)` genes of `r ncol(sce)` cells.


## Data exploration and dimensionality reduction

- Visualisation is a first essential step to learn from data


```{r}
counts(sce)[18990:19000,9:20]
```

- It is impossible to learn about the structure in the data by staring at the data matrix.
- We should be able to explore the data in a low dimensional projection

```{r}
plotReducedDim(sce, dimred="TSNE", colour_by="cell")
plotReducedDim(sce, dimred="TSNE", colour_by="stim")
```

- Note, that we see huge effect of treatment. If I see this I am always on my guard!
- We contacted the authors and learned that all control cells were sequenced in a first run and all stimulated cells were on a second sequencing run. So the large effect might be an effect of batch!

## Prediction

```{r}
plotReducedDim(sce, dimred="TSNE", colour_by="cell")
```

- In single cell analysis it is key to identify cell types in scRNA-seq data sets before in-depth investigations of their functional and pathological roles.

- Use models that were build based on reference data sets to predict cell types in new data sets based on their gene expression pattern

- Problem: we have `r nrow(sce)` genes at our disposal to build this prediction model!

- Other examples

  - Prediction of risk on mortality is used on a daily basis in intensive care units to prioritise patient care.

  - Facebook: predict the identity of the faces of people that are on an new image that is uploaded.

  - Netflix: Suggest movies that you would like

## Large scale hypothesis testing

```{r}
plotReducedDim(sce, dimred="TSNE", colour_by="cell", shape_by="stim")
```

- Which genes are differentially expressed between control and stimulated treatment  (assess in each cell type)
- For which genes we see that the differential expression according to treatment is changing according to the cell type (interaction)
- We have to model the gene expression for each gene

$$
\left \{
\begin{array}{lcl}
y_{ig} &\sim& NB(\mu_{ig}, \phi_{g})\\
E[y_{ig}] &=& \mu\\
\log(\mu_{ig}) &=& \eta_{ig}\\
\eta_{ig} &=& \beta_{0,g} + \sum\limits_{c=1}^C \beta_{c,g} X_{ic} + \beta_{s,g} X_{is} + \sum\limits_{c=1}^C \beta_{c:s,g} X_{ic} X_{is} + \alpha_{ig}
\end{array} \right.
$$

With

- Cell type $c=2\ldots C$ and $X_{ic}$ is a dummy variable that $X_{ic}=1$ if cell $i$ is of cell type c and $X_{ic}=0$ otherwise. Note that cell type $c=1$ is the reference cell type

- Indicator $X_{is}$ indicates if cell $i$ was stimulated $X_{is}=1$ or not $X_{is}=0$. So the control treatment is the reference treatment.

Suppose we want to test if the effect of the stimulus (average difference in expression between stimulated and non stimulated cells) is different in cell type c than in the reference cell type 1?

- $H_0: \beta_{c:s,g} = 0$
- $H_1: \beta_{c:s,g} \neq 0$


- We have to assess this for `r nrow(sce)` genes!
- If we assess each test at the 5% level we can expect 0.05 * `r nrow(sce)` = `r round(0.05*nrow(sce),0)` false positives.

$\rightarrow$ massive multiple testing problem!


Note, that we cannot differentiate between batch and treatment because of the flaw in the experimental design!

Other examples

- Find regions (voxels) in the brain (on brain image) that are associated with a certain condition/treatment
- Evaluation of trading rules



---


# Linear regression

- Linear regression is a very important statistical tool to study the association between variables and to build prediction models.

## Scalar form

- Consider a vector of predictors $\mathbf{x}=(x_1,\ldots,x_p)$ and
- a real-valued response $Y$
- then the linear regression model can be written as
\[
Y=f(\mathbf{x}) +\epsilon=\beta_0+\sum\limits_{j=1}^p x_j\beta_j + \epsilon
\]
with i.i.d. $\epsilon\sim N(0,\sigma^2)$
- We will typically work on mean centered variables: $Y-\bar{Y}$ and $X - \bar{X}$, then the intercept is dropped from the model:
\[
Y=f(\mathbf{x}) +\epsilon=\sum\limits_{j=1}^p x_j\beta_j + \epsilon
\]


## Vector/Matrix form

- $n$ observations $(\mathbf{x}_1,y_1) \ldots (\mathbf{x}_n,y_n)$ with $\mathbf{x}_1^T=[1 x_1 \ldots x_p]$
- Regression in matrix notation
\[\mathbf{Y}=\mathbf{X\beta} + \mathbf{\epsilon}\]
with $\mathbf{Y}=\left[\begin{array}{c}y_1\\ \vdots\\y_n\end{array}\right]$,
$\mathbf{X}=\left[\begin{array}{cccc} 1&x_{11}&\ldots&x_{1p}\\
\vdots&\vdots&&\vdots\\
1&x_{n1}&\ldots&x_{np}
\end{array}\right]$ or $\mathbf{X}=\left[\begin{array}{c} \mathbf{x}_1^T\\\vdots\\\mathbf{x}_n^T\end{array}\right]$,
$\boldsymbol{\beta}=\left[\begin{array}{c}\beta_0\\ \vdots\\ \beta_p\end{array}\right]$ and
$\mathbf{\epsilon}=\left[\begin{array}{c} \epsilon_1 \\ \vdots \\ \epsilon_n\end{array}\right]$

Note, that upon centering of $\mathbf{X}$ and $\mathbf{Y}$ the 1 is dropped from each $\mathbf{x}_i$ and thus the column of 1 is dropped in $\mathbf{X}$

---

## Interpretation

From the linear regression we get
\[
  E[Y \mid \mathbf{x}] = \mathbf{x}^T\boldsymbol{\beta} .
\]

- Hence, the $\beta$ parameters relate the regressor $\mathbf{x}$ to the mean outcome.
- If we know the covariate pattern $\mathbf{x}$ we can use the model to predict $Y$.

Upon centering we get the following for a model with a single regressor
\[
  E[Y \mid x] = \beta_1 x
\]
and
\[
 \beta_1 = E[Y\mid x+1] - E[Y\mid x] .
\]

Hence, the $\beta_1$ parameter has an interpretation as the mean increase of the outcome as the regressor increases with one unit. The parameter does not say very about individual outcomes. It is the residual variance $\sigma^2$ that determines how much individual outcomes vary about the mean outcome.

The $\beta$ parameters are used to measure association, but a $\beta \neq 0$ does not necessarily mean that the model will give good predictions.

---

- In a later chapter we will discuss the problem of large scale hypothesis testing: testing many hypotheses in a single study (ten to hundred thousands of hypotheses).

- A statistical test is constructed to control the type I error rate at the significance level. However, when many hypotheses are to be tested in a single study, the probability to find false associations is no longer controlled if p-values are compared to the significance level.

- We will later introduce the concept of false discovery rates to overcome the problem.

## Least Squares (LS)

- Minimize the residual sum of squares
\begin{eqnarray*}
RSS(\boldsymbol{\beta})&=&\sum\limits_{i=1}^n e^2_i\\
&=&\sum\limits_{i=1}^n \left(y_i-\beta_0-\sum\limits_{j=1}^p x_{ij}\beta_j\right)^2
\end{eqnarray*}

- or in matrix notation
\begin{eqnarray*}
RSS(\boldsymbol{\beta})&=&(\mathbf{Y}-\mathbf{X\beta})^T(\mathbf{Y}-\mathbf{X\beta})\\
&=&\Vert \mathbf{Y}-\mathbf{X\beta}\Vert^2_2
\end{eqnarray*}
with the $L_2$-norm of a $p$-dim. vector $v$ $\Vert \mathbf{v} \Vert_2=\sqrt{v_1^2+\ldots+v_p^2}$
$\rightarrow$ $\hat{\boldsymbol{\beta}}=\text{argmin}_\beta \Vert \mathbf{Y}-\mathbf{X\beta}\Vert^2$

---

### Minimize RSS
\[
\begin{array}{ccc}
\frac{\partial RSS}{\partial \boldsymbol{\beta}}&=&\mathbf{0}\\\\
\frac{(\mathbf{Y}-\mathbf{X\beta})^T(\mathbf{Y}-\mathbf{X}\boldsymbol{\beta})}{\partial \boldsymbol{\beta}}&=&\mathbf{0}\\\\
-2\mathbf{X}^T(\mathbf{Y}-\mathbf{X}\boldsymbol{\beta})&=&\mathbf{0}\\\\
\mathbf{X}^T\mathbf{X\beta}&=&\mathbf{X}^T\mathbf{Y}\\\\
\hat{\boldsymbol{\beta}}&=&(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{Y}
\end{array}
\]

It can be shown that the estimator is unbiased:

$$
E[\hat{\boldsymbol{\beta}}]=\boldsymbol{\beta}
$$

---

### Geometrical Interpretation

Toy Example: fit without intercept

- n=3 and p=2
\[
\mathbf{X}=\left[\begin{array}{cc}
2&0\\
0&2\\
0&0
\end{array}\right]
\]

```{r}
set.seed(4)
x1 <- c(2, 0, 0)
x2 <- c(0, 2, 0)
y <- x1 * 0.5 + x2 * 0.5 + rnorm(3, 2)
fit <- lm(y ~ -1 + x1 + x2)
```

- $\mathbf{Y}$ is now given by

\[
\mathbf{Y}=\left[\begin{array}{c}
1 + \epsilon_1 \\
1 + \epsilon_2\\
0 + \epsilon_3
\end{array}\right]
\]

with the $\epsilon_i$s some random noise.

- The actual values of $\mathbf{Y}$ are:

```{r}
## True Y
y
## Fitted Y
fit$fitted.values
```

#### Visualise fit

```{r}
# predict values on regular xy grid
x1pred <- seq(-1, 3, length.out = 10)
x2pred <- seq(-1, 3, length.out = 10)
xy <- expand.grid(x1 = x1pred,
x2 = x2pred)
ypred <- matrix (nrow = 30, ncol = 30,
data = predict(fit, newdata = data.frame(xy),
interval = "prediction"))

library(plot3D)


# fitted points for droplines to surface
th=20
ph=5
scatter3D(x1,
  x2,
  y,
  pch = 16,
  col="darkblue",
  cex = 1,
  theta = th,
  ticktype = "detailed",
  xlab = "x1",
  ylab = "x2",
  zlab = "y",
  colvar=FALSE,
  bty = "g",
  xlim=c(-1,3),
  ylim=c(-1,3),
  zlim=c(-2,4))

for (i in 1:3)
  lines3D(
    x = rep(x1[i],2),
    y = rep(x2[i],2),
    z = c(y[i],fit$fitted[i]),
    col="darkblue",
    add=TRUE,
    lty=2)

z.pred3D <- outer(
  x1pred,
  x2pred,
  function(x1,x2)
  {
    fit$coef[1]*x1+fit$coef[2]*x2
  })

x.pred3D <- outer(
  x1pred,
  x2pred,
  function(x,y) x)

y.pred3D <- outer(
  x1pred,
  x2pred,
  function(x,y) y)

surf3D(
  x.pred3D,
  y.pred3D,
  z.pred3D,
  col="blue",
  facets=NA,
  add=TRUE)
```

#### Projection

- We can also interpret the fit as the projection of the $n\times 1$ vector $\mathbf{Y}$ on the column space of the matrix $\mathbf{X}$.

- So each column in $\mathbf{X}$ is also an $n\times 1$ vector.

- For the toy example n=3 and p=2.
So the column space of X is a plane in the three dimensional space.

\[
\hat{\mathbf{Y}} = \mathbf{X} (\mathbf{X}^T\mathbf{X})^{-1} \mathbf{X}^T \mathbf{Y}
\]

1. Plane spanned by column space:
```{r}
arrows3D(
  0, 0, 0, x1[1], x1[2], x1[3],
  xlim = c(0, 5), ylim = c(0, 5), zlim = c(0, 5), bty = "g",
  theta = th, col = 2, ticktype = "detailed",
  xlab = "observation 1", ylab = "observation 2", zlab = "observation 3"
)
text3D(x1[1],x1[2],x1[3],labels="X1",col=2,add=TRUE)
arrows3D(0,0,0,x2[1],x2[2],x2[3],add=TRUE,col=2)
text3D(x2[1],x2[2],x2[3],labels="X2",col=2,add=TRUE)
```

2. Vector of Y:

Actual values of $\mathbf{Y}$:

```{r}
y
```

\[
\mathbf{Y}=\left[\begin{array}{c}
`r y[1]` \\
`r y[2]` \\
`r y[3]`
\end{array}\right]
\]

```{r}
arrows3D(
  0, 0, 0, x1[1], x1[2], x1[3],
  xlim = c(0, 5), ylim = c(0, 5), zlim = c(0, 5), bty = "g",
  theta = th, col = 2, ticktype = "detailed",
  xlab = "observation 1", ylab = "observation 2", zlab = "observation 3"
)
text3D(x1[1],x1[2],x1[3],labels="X1",col=2,add=TRUE)
arrows3D(0,0,0,x2[1],x2[2],x2[3],add=TRUE,col=2)
text3D(x2[1],x2[2],x2[3],labels="X2",col=2,add=TRUE)
arrows3D(0,0,0,y[1],y[2],y[3],add=TRUE,col="darkblue")
text3D(y[1],y[2],y[3],labels="Y",col="darkblue",add=TRUE)
```

3. Projection of Y onto column space

Actual values of fitted values $\mathbf{\hat{Y}}$:

```{r}
fitted.values(fit)
```

\[
\mathbf{Y}=\left[\begin{array}{c}
`r fitted.values(fit)[1]` \\
`r fitted.values(fit)[2]` \\
`r fitted.values(fit)[3]`
\end{array}\right]
\]

```{r}
arrows3D(
  0, 0, 0, x1[1], x1[2], x1[3],
  xlim = c(0, 5), ylim = c(0, 5), zlim = c(0, 5), bty = "g",
  theta = th, col = 2, ticktype = "detailed",
  xlab = "observation 1", ylab = "observation 2", zlab = "observation 3"
)
text3D(x1[1],x1[2],x1[3],labels="X1",col=2,add=TRUE)
arrows3D(0,0,0,x2[1],x2[2],x2[3],add=TRUE,col=2)
text3D(x2[1],x2[2],x2[3],labels="X2",col=2,add=TRUE)
arrows3D(0,0,0,y[1],y[2],y[3],add=TRUE,col="darkblue")
text3D(y[1],y[2],y[3],labels="Y",col="darkblue",add=TRUE)
arrows3D(0,0,0,fit$fitted[1],fit$fitted[2],fit$fitted[3],add=TRUE,col="darkblue")
segments3D(y[1],y[2],y[3],fit$fitted[1],fit$fitted[2],fit$fitted[3],add=TRUE,lty=2,col="darkblue")
text3D(fit$fitted[1],fit$fitted[2],fit$fitted[3],labels="fit",col="darkblue",add=TRUE)
```

- Note, that it is also clear from the equation in the derivation of the LS that the residual is orthogonal on the column space:
\[
 -2 \mathbf{X}^T(\mathbf{Y}-\mathbf{X}\boldsymbol{\beta}) = 0
\]

- Imagine what happens when p >> n!!!

$\rightarrow$ curse of dimensionality!

## Variance Estimator?
\[
\begin{array}{ccl}
\hat{\boldsymbol{\Sigma}}_{\hat{\boldsymbol{\beta}}}
&=&\text{var}\left[(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{Y}\right]\\\\
&=&(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\text{var}\left[\mathbf{Y}\right]\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}\\\\
&=&(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T(\mathbf{I}\sigma^2)\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}
\\\\
&=&(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{I}\quad\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2\\\\
%\hat{\boldmath{\Sigma}}_{\hat{\boldsymbol{\beta}}}&=&(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T \text{var}\left[\mathbf{Y}\right](\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}\\
&=&(\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{X}(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2\\\\
&=&(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2
\end{array}
\]

The fact that $\hat{\boldsymbol{\beta}}$ is unbiased and has a variance of $(\mathbf{X}^T\mathbf{X})^{-1}\sigma^2$ will be important when assessing association!

---

## Prediction error

Least squares estimators are unbiased and consistent, but these properties are not very important for prediction models.

A prediction model is considered good if it can predict well outcomes.

The **prediction error** for a prediction at predictor $\mathbf{x}$ is given by

$$
   \hat{Y}(\mathbf{x}) - Y^*,
$$

where

-  $\hat{Y}(\mathbf{x})=\mathbf{x}^T\hat{\boldsymbol{\beta}}$ is the prediction at $\mathbf{x}$

-  $Y^*$ is an outcome at predictor $\mathbf{x}$

 Since prediction is typically used to predict an outcome before it is observed, the outcome $Y^*$ is not observed yet. Hence, the prediction error cannot be computed.

 ---

 The problem of unobservable prediction errors is partly solved by the **expected conditional test error** (sometimes referred to as the mean squared error, MSE)
\[
  \text{Err}(\mathbf{x})=E[(\hat{Y}(\mathbf{x}) - Y^*)^2].
\]

With (suppressing the dependence on $\mathbf{x}$)
\[
  \mu = E[\hat{Y}] \text{ and } \mu^*=E[Y^*]
\]
the error can be expressed as
\begin{eqnarray*}
  \text{Err}
    &=&E\left\{\left[(\hat{Y} -\mu)-(Y-\mu^*)-(\mu^*-\mu)\right]^2\right\} \\
    &=& E[(\hat{Y} -\mu)^2]+E[(Y-\mu^*)^2]+E[(\mu^*-\mu)^2] \\
    &=& \text{var}[\hat{Y}] +\text{var}[Y] +\text{bias}^2
\end{eqnarray*}

 The term $\text{var}[Y]$ (irreducible error) does not depend on the model and may therefore be ignored when $\text{Err}$ is used for comparing prediction models.

 ---


In this introductory chapter we only aim to give a rough discussion on prediction errors. Later definitions will be refined and the notation will be more accurate. Also a more detailed discussion on the bias-variance trade-off will follow. For the moment it is sufficient to vaguely know that:

- the expected conditional test error is introduced to circumvent the problem that the prediction error cannot be observed. In later chapters we will look at estimators of the expected error.

- the expected conditional test error  is in some literature also known as the *mean squared error* (MSE), but we do not adopt this terminology because MSE is also commonly used to refer to SSE divided by the residual degrees of freedom in a linear regression model.

- The identity $\text{Err}=\text{var}[\hat{Y}] +\text{var}[Y] +\text{bias}^2$ is known as the bias-variance trade-off. It shows that a good prediction model (i.e. a model resulting in a small Err), can be obtained by a model that shows a small bias as long as this bias is compensated with a large reduction of the variance or the predictions. A more detailed discussion will follow in later chapters.

- For prediction models with a large number of predictors we will therefore introduce penalized regression. This will induce some bias in the estimation, but will allow us to reduce the variance considerably.

# Session info {-}

<details><summary>Session info</summary>

```{r session_info, echo=FALSE, cache=FALSE}
Sys.time()
sessioninfo::session_info()
```

</details>
</div>
<div class="footer">
    <hr>
    This work is licensed under the <a href= "https://creativecommons.org/licenses/by-nc-sa/4.0">
    CC BY-NC-SA 4.0</a> licence.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("intro.Rmd");
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
