<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lieven Clement" />


<title>Linear Discriminant Analysis (LDA)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>


<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">HDDA21</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Lectures
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="intro.html">1. Introduction</a>
    </li>
    <li>
      <a href="svd.html">2. Singular Value Decomposition</a>
    </li>
    <li>
      <a href="svdGeometricInterpretation.html">2.3. Geometric Interpretation SVD</a>
    </li>
    <li>
      <a href="MDS_linkGramDistanceMatrix.html">2.7. Link MDS and Gram Distance Matrix</a>
    </li>
    <li>
      <a href="prediction.html">3. Prediction with High Dimensional Predictors</a>
    </li>
    <li>
      <a href="sparseSvd.html">4. Sparse Singular Value Decomposition</a>
    </li>
    <li>
      <a href="lda.html">5. Linear Discriminant Analysis</a>
    </li>
    <li>
      <a href="lsi.html">6. Large Scale Inference</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-laptop"></span>
     
    Labs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Coming soon!</li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/statOmics/HDDA21">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://statomics.github.io/">statOmics</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
<li role="separator" class="divider"></li>
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Linear Discriminant Analysis (LDA)</h1>
<h4 class="author">Lieven Clement</h4>
<h4 class="date">statOmics, Ghent University (<a href="https://statomics.github.io" class="uri">https://statomics.github.io</a>)</h4>

</div>


<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>## ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
<pre><code>## ✔ ggplot2 3.3.5     ✔ purrr   0.3.4
## ✔ tibble  3.1.5     ✔ dplyr   1.0.7
## ✔ tidyr   1.1.4     ✔ stringr 1.4.0
## ✔ readr   2.0.1     ✔ forcats 0.5.1</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<pre class="r"><code>library(gridExtra)</code></pre>
<pre><code>## 
## Attaching package: &#39;gridExtra&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:dplyr&#39;:
## 
##     combine</code></pre>
<div id="breast-cancer-example" class="section level1">
<h1><span class="header-section-number">1</span> Breast cancer example</h1>
<ul>
<li><p>Schmidt <em>et al.</em>, 2008, Cancer Research, <strong>68</strong>, 5405-5413</p></li>
<li><p>Gene expression patterns in n=200 breast tumors were investigated (p=22283 genes)</p></li>
<li><p>After surgery the tumors were graded by a pathologist (stage 1,2,3)</p></li>
</ul>
<div id="data" class="section level2">
<h2><span class="header-section-number">1.1</span> Data</h2>
<pre class="r"><code>#BiocManager::install(&quot;genefu&quot;)
#BiocManager::install(&quot;breastCancerMAINZ&quot;)

library(genefu)
library(breastCancerMAINZ)
data(mainz)

X &lt;- t(exprs(mainz)) # gene expressions
n &lt;- nrow(X)
H &lt;- diag(n)-1/n*matrix(1,ncol=n,nrow=n)
X &lt;- H%*%X
Y &lt;- pData(mainz)$grade
table(Y)</code></pre>
<pre><code>## Y
##   1   2   3 
##  29 136  35</code></pre>
<pre class="r"><code>svdX &lt;- svd(X)
k &lt;- 2
Zk &lt;- svdX$u[,1:k] %*% diag(svdX$d[1:k])
colnames(Zk) &lt;- paste0(&quot;Z&quot;,1:k)

Zk %&gt;%
  as.data.frame %&gt;%
  mutate(grade = Y %&gt;% as.factor) %&gt;%
  ggplot(aes(x= Z1, y = Z2, color = grade)) +
  geom_point(size = 3)</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
</div>
<div id="linear-discriminant-analysis" class="section level1">
<h1><span class="header-section-number">2</span> Linear discriminant analysis</h1>
<p>Fisher’s construction of LDA is simple: it allows for classification in a dimension-reduced subspace of <span class="math inline">\(\mathbb{R}^p\)</span>.</p>
<p>First we assume that <span class="math inline">\(Y\)</span> can only take two values (0/1).</p>
<p>Fisher aimed for a direction, say <span class="math inline">\(\mathbf{a}\)</span>, in the <span class="math inline">\(p\)</span>-dimensional predictor space such that the orthogonal projections of the predictors, <span class="math inline">\(\mathbf{x}^t\mathbf{a}\)</span>, show maximal ratio between the between and within sums of squares: <span class="math display">\[
  \mathbf{v} = \text{ArgMax}_a \frac{\mathbf{a}^t\mathbf{B}\mathbf{a}}{\mathbf{a}^t\mathbf{W}\mathbf{a}} \text{ subject to } {\mathbf{a}^t\mathbf{W}\mathbf{a}}=1,
 \]</span> where <span class="math inline">\(\mathbf{W}\)</span> and <span class="math inline">\(\mathbf{B}\)</span> are the within and between covariance matrices of <span class="math inline">\(\mathbf{x}\)</span>. The restriction is introduced to obtain a (convenient) unique solution.</p>
<hr />
<div id="between-and-within-sums-of-squares" class="section level2">
<h2><span class="header-section-number">2.1</span> Between and within sums of squares</h2>
<ul>
<li><p>In the training dataset, let <span class="math inline">\(\mathbf{x}_{ik}\)</span> denote the <span class="math inline">\(i\)</span>th <span class="math inline">\(p\)</span>-dimensional observation in the <span class="math inline">\(k\)</span>th group (<span class="math inline">\(k=0,1\)</span> referring to <span class="math inline">\(Y=0\)</span> and <span class="math inline">\(Y=1\)</span>, resp.), <span class="math inline">\(i=1,\ldots, n_k\)</span>.</p></li>
<li><p>Let <span class="math inline">\(z_{ik}=\mathbf{a}^t\mathbf{x}_{ik}\)</span> denote the orthogonal projection of <span class="math inline">\(\mathbf{x}_{ik}\)</span> onto <span class="math inline">\(\mathbf{a}\)</span></p></li>
<li><p>For the one-dimensional <span class="math inline">\(z\)</span>-observations, consider the following sum of squares: <span class="math display">\[
  \text{SSE}=\text{within sum of squares} = \sum_{k=0,1}\sum_{i=1}^{n_k} (z_{ik}-\bar{z}_k)^2
\]</span> <span class="math display">\[
  \text{SSB}=\text{between sum of squares} = \sum_{k=0,1}\sum_{i=1}^{n_k} (\bar{z}_{k}-\bar{z})^2 = \sum_{k=0,1} n_k (\bar{z}_{k}-\bar{z})^2
\]</span> with <span class="math inline">\(\bar{z}_k\)</span> the sample mean of <span class="math inline">\(z_{ik}\)</span> within group <span class="math inline">\(k\)</span>, and <span class="math inline">\(\bar{z}\)</span> the sample mean of all <span class="math inline">\(z_{ik}\)</span>.</p></li>
<li><p>To reformulate SSE and SSB in terms of the <span class="math inline">\(p\)</span>-dimensional <span class="math inline">\(\mathbf{x}_{ik}\)</span>, we need the sample means <span class="math display">\[
  \bar{z}_k = \frac{1}{n_k} \sum_{i=1}^{n_k} z_{ik} = \frac{1}{n_k} \sum_{i=1}^{n_k} \mathbf{a}^t\mathbf{x}_{ik} = \mathbf{a}^t \frac{1}{n_k} \sum_{i=1}^{n_k} \mathbf{x}_{ik} = \mathbf{a}^t \bar{\mathbf{x}}_k
\]</span> <span class="math display">\[
  \bar{z} = \frac{1}{n}\sum_{k=0,1}\sum_{i=1}^{n_k} z_{ik} = \cdots = \mathbf{a}^t\bar{\mathbf{x}}.
\]</span></p></li>
<li><p>SSE becomes <span class="math display">\[
  \text{SSE} = \sum_{k=0,1}\sum_{i=1}^{n_k} (z_{ik}-\bar{z}_k)^2 = \mathbf{a}^t \left(\sum_{k=0,1}\sum_{i=1}^{n_k} (\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)(\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)^t\right)\mathbf{a}
\]</span></p></li>
<li><p>SSB becomes <span class="math display">\[
  \text{SSB} = \sum_{k=0,1} n_k (\bar{z}_{k}-\bar{z})^2 = \mathbf{a}^t \left(\sum_{k=0,1} n_k (\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})(\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})^t \right)\mathbf{a}
\]</span></p></li>
</ul>
<hr />
<ul>
<li><p>The <span class="math inline">\(p \times p\)</span> matrix <span class="math display">\[
  \mathbf{W}=\sum_{k=0,1}\sum_{i=1}^{n_k} (\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)(\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)^t
 \]</span> is referred to as the matrix of <strong>within</strong> sum of squares and cross products.</p></li>
<li><p>The <span class="math inline">\(p \times p\)</span> matrix <span class="math display">\[
  \mathbf{B}=\sum_{k=0,1} n_k (\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})(\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})^t
 \]</span> is referred to as the matrix of <strong>between</strong> sum of squares and cross products.</p></li>
<li><p>Note that on the diagonal of <span class="math inline">\(\mathbf{W}\)</span> and <span class="math inline">\(\mathbf{B}\)</span> you find the ordinary univariate within and between sums of squares of the individual components of <span class="math inline">\(\mathbf{x}\)</span>.</p></li>
</ul>
</div>
<div id="obtain-projections" class="section level2">
<h2><span class="header-section-number">2.2</span> Obtain projections</h2>
<p>An equivalent formulation: <span class="math display">\[
  \mathbf{v} = \text{ArgMax}_a \mathbf{a}^t\mathbf{B}\mathbf{a} \text{ subject to } \mathbf{a}^t\mathbf{W}\mathbf{a}=1.
 \]</span></p>
<p>This can be solved by introducing a Langrange multiplier: <span class="math display">\[
  \mathbf{v} = \text{ArgMax}_a \mathbf{a}^t\mathbf{B}\mathbf{a} -\lambda(\mathbf{a}^t\mathbf{W}\mathbf{a}-1).
 \]</span></p>
<p>Calculating the partial derivative w.r.t. <span class="math inline">\(\mathbf{a}\)</span> and setting it to zero gives <span class="math display">\[\begin{eqnarray*}
   2\mathbf{B}\mathbf{a} -2\lambda \mathbf{W}\mathbf{a} &amp;=&amp; 0\\
   \mathbf{B}\mathbf{a}  &amp;=&amp; \lambda \mathbf{W}\mathbf{a} \\
   \mathbf{W}^{-1}\mathbf{B}\mathbf{a} &amp;=&amp; \lambda\mathbf{a}.
 \end{eqnarray*}\]</span></p>
<p>From the final equation we recognise that <span class="math inline">\(\mathbf{v}=\mathbf{a}\)</span> is an <strong>eigenvector</strong> of <span class="math inline">\(\mathbf{W}^{-1}\mathbf{B}\)</span>, and <span class="math inline">\(\lambda\)</span> is the corresponding <strong>eigenvalue</strong>.</p>
<p>The equation has in general <span class="math inline">\(\text{rank}(\mathbf{W}^{-1}\mathbf{B})\)</span> solutions. In the case of two classes, the rank equals 1 and thus only one solution exists.</p>
<hr />
<ul>
<li><p>A training data set is used for the calculation of <span class="math inline">\(\mathbf{W}\)</span> and <span class="math inline">\(\mathbf{B}\)</span>. <span class="math inline">\(\longrightarrow\)</span> This gives the eigenvector <span class="math inline">\(\mathbf{v}\)</span></p></li>
<li><p>The training data is also used for the calculation of the centroids of the classes (e.g. the sample means, say <span class="math inline">\(\bar{\mathbf{x}}_1\)</span> and <span class="math inline">\(\bar{\mathbf{x}}_2\)</span>). <span class="math inline">\(\longrightarrow\)</span> The projected centroids are given by <span class="math inline">\(\bar{\mathbf{x}}_1^t\mathbf{v}\)</span> and <span class="math inline">\(\bar{\mathbf{x}}_2^t\mathbf{v}\)</span>.</p></li>
<li><p>A new observation with predictor <span class="math inline">\(\mathbf{x}\)</span> is classified in the class for which the projected centroid is closest to the projected predictor <span class="math inline">\(z=\mathbf{x}^t\mathbf{v}\)</span>.</p></li>
</ul>
<p>An advantage of this approach is that <span class="math inline">\(\mathbf{v}\)</span> can be interpreted (similar as the loadings in a PCA) in terms of which predictors <span class="math inline">\(x_j\)</span> are important to discriminate between classes 0 and 1.</p>
<hr />
</div>
<div id="more-than-two-classes" class="section level2">
<h2><span class="header-section-number">2.3</span> More than two classes</h2>
<p>When the outcome <span class="math inline">\(Y\)</span> refers to more than two classes, say <span class="math inline">\(m\)</span> classes, then Fisher’s method is constructed in exactly the same way. Now <span class="math display">\[
  \mathbf{W}^{-1}\mathbf{B}\mathbf{a} = \lambda\mathbf{a}
\]</span> will have <span class="math inline">\(r=\text{rank}(\mathbf{W}^{-1}\mathbf{B}) = \min(m-1,p,n)\)</span> solutions (eigenvectors and eigenvalues). (<span class="math inline">\(n\)</span>: sample size of training data)</p>
<p>Let <span class="math inline">\(\mathbf{v}_j\)</span> and <span class="math inline">\(\lambda_j\)</span> denote the <span class="math inline">\(r\)</span> solutions, and define</p>
<ul>
<li><p><span class="math inline">\(\mathbf{V}\)</span>: <span class="math inline">\(p\times r\)</span> matrix with collums <span class="math inline">\(\mathbf{v}\)</span></p></li>
<li><p><span class="math inline">\(\mathbf{L}\)</span>: <span class="math inline">\(r \times r\)</span> diagonal matrix with elements <span class="math inline">\(\lambda_1 &gt; \lambda_2 &gt; \cdots &gt; \lambda_r\)</span></p></li>
</ul>
<p>The <span class="math inline">\(p\)</span>-dimensional predictor data in <span class="math inline">\(\mathbf{X}\)</span> may then be transformed to the <span class="math inline">\(r\)</span>-dimensional scores <span class="math display">\[
  \mathbf{Z} = \mathbf{X}\mathbf{V}.
\]</span></p>
<hr />
<p>For eigenvectors <span class="math inline">\(\mathbf{v}_i\)</span> and <span class="math inline">\(\mathbf{v}_j\)</span>, it holds that <span class="math display">\[
 \text{cov}\left[Z_i,Z_j\right] = \text{cov}\left[\mathbf{X}\mathbf{v}_i,\mathbf{X}\mathbf{v}_j\right]= \mathbf{v}_i^t \mathbf{W} \mathbf{v}_j = \delta_{ij} ,
\]</span> in which the covariances are defined within groups. Hence, within the groups (classes) the scores are uncorrelated.</p>
</div>
</div>
<div id="high-dimensional-predictors" class="section level1">
<h1><span class="header-section-number">3</span> High dimensional predictors</h1>
<p>With high-dimensional predictors</p>
<ul>
<li><p>Replace the <span class="math inline">\(p\times p\)</span> matrices <span class="math inline">\(\mathbf{W}\)</span> and <span class="math inline">\(\mathbf{B}\)</span> by their diagonal matrices (i.e. put zeroes on the off-diagonal positions)</p></li>
<li><p><strong>Sparse LDA</strong> by imposing an <span class="math inline">\(L_1\)</span>-penalty on <span class="math inline">\(\mathbf{v}\)</span>.</p></li>
</ul>
<p>Two approaches: Zhou <em>et al.</em> (2006), Journal of Computational and Graphical Statistics , <strong>15</strong>, 265-286, and <a href="https://web.stanford.edu/~hastie/Papers/sda_resubm_daniela-final.pdf">Clemmensen <em>et al.</em> (2011), Technometrics, <strong>53</strong></a>.</p>
</div>
<div id="breast-cancer-example-1" class="section level1">
<h1><span class="header-section-number">4</span> Breast cancer example</h1>
<div id="all-genes" class="section level2">
<h2><span class="header-section-number">4.1</span> All genes</h2>
<div id="lda" class="section level3">
<h3><span class="header-section-number">4.1.1</span> LDA</h3>
<ul>
<li><p>Fisher’s LDA is illustrated on the breast cancer data with all three tumor stages as outcome.</p></li>
<li><p>We try to discriminate between the different stages according to the gene expression data of all genes.</p></li>
<li><p>We cache the result because the calculation takes 10 minutes.</p></li>
</ul>
<pre class="r"><code>breast.lda &lt;- MASS::lda(x = X, grouping = Y)</code></pre>
<pre><code>## Warning in lda.default(x, grouping, ...): variables are collinear</code></pre>
<pre class="r"><code>Vlda &lt;- breast.lda$scaling
colnames(Vlda) &lt;- paste0(&quot;V&quot;,1:ncol(Vlda))

Zlda &lt;- X%*%Vlda
colnames(Zlda) &lt;- paste0(&quot;Z&quot;,1:ncol(Zlda))

grid.arrange(
  Zlda %&gt;%
    as.data.frame %&gt;%
    mutate(grade = Y %&gt;% as.factor) %&gt;%
    ggplot(aes(x= Z1, y = Z2, color = grade)) +
    geom_point(size = 3) +
    coord_fixed(),
  ggplot() +
    geom_bar(aes(x = c(&quot;z1&quot;,&quot;z2&quot;), y = breast.lda$svd), stat = &quot;identity&quot;) +
    xlab(&quot;Discriminant&quot;) +
    ylab(&quot;Eigen Values&quot;),
  layout_matrix = matrix(
    c(1,1,2),
    nrow=1)
  )</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<hr />
<ul>
<li><p>The columns of the matrix <span class="math inline">\(\mathbf{V}\)</span> contain the eigenvectors. There are <span class="math inline">\(\min(3-2,22283,200)=2\)</span> eigenvectors. The <span class="math inline">\(200\times 2\)</span> matrix <span class="math inline">\(\mathbf{Z}\)</span> contains the scores on the two Fisher discriminants.</p></li>
<li><p>The eigenvalue <span class="math inline">\(\lambda_j\)</span> can be interpreted as the ratio <span class="math display">\[
\frac{\mathbf{v}_j^t\mathbf{B}\mathbf{v}_j}{\mathbf{v}_j^t\mathbf{W}\mathbf{v}_j} ,
\]</span> or (upon using <span class="math inline">\(\mathbf{v}_j^t\mathbf{W}\mathbf{v}_j=1\)</span>) the between-centroid sum of squares (in the reduced dimension space of the Fisher discriminants) <span class="math display">\[
\mathbf{v}_j^t\mathbf{B}\mathbf{v}_j.
\]</span></p></li>
<li><p>From the screeplot of the eigenvalues we see that the first dimension is more important than the second (not hugely) in terms of discriminating between the groups.</p></li>
<li><p>From the scatterplot we can see that there is no perfect separation (discrimination) between the three tumor stages (quite some overlap).</p></li>
<li><p>To some extent the first Fisher discriminant dimension discriminates stage 3 (green dots) from the other two stages, and the second dimension separates stage 1 (black dots) from the two others.</p></li>
</ul>
</div>
<div id="interpretation-of-loadings" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Interpretation of loadings</h3>
<pre class="r"><code>grid.arrange(
  Vlda %&gt;%
    as.data.frame %&gt;%
    mutate(geneID = 1:nrow(Vlda)) %&gt;%
    ggplot(aes(x = geneID, y = V1)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,1]), col = &quot;red&quot;),

  Vlda %&gt;%
    as.data.frame %&gt;%
    mutate(geneID = 1:nrow(Vlda)) %&gt;%
    ggplot(aes(x = geneID, y = V2)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,2]), col = &quot;red&quot;),
  ncol = 2)</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>The loadings of the Fisher discriminants are within the columns of the <span class="math inline">\(\mathbf{V}\)</span> matrix.</p>
<ul>
<li><p>Since we have 22283 genes, each discriminant is a linear combination of 22283 gene expression. Instead of looking at the listing of 22283 loadings, we made an index plot (no particular ordering of genes on horizontal axis).</p></li>
<li><p>The red horizontal reference lines correspond to the average of the loading (close to zero) and the average plus and minus twice the standard deviation of the loadings.</p></li>
<li><p>If no genes had any “significant” discriminating power, then we would expect approximately <span class="math inline">\(95\%\)</span> of all loadings within the band. Thus loadings outside of the band are of potential interest and may perhaps be discriminating between the three tumor stages.</p></li>
<li><p>In the graphs presented here we see many loadings within the bands, but also many outside of the band.</p></li>
</ul>
<p>We repeat the analysis, but now with the sparse LDA method of Clemmensen et al. (2011).</p>
<hr />
</div>
</div>
<div id="sparse-lda-based-on-150-random-genes" class="section level2">
<h2><span class="header-section-number">4.2</span> Sparse LDA based on 150 random genes</h2>
<ul>
<li><p>We only present the results of the sparse LDA based on a random subset of 150 genes (ordering of genes in datamatrix is random).</p></li>
<li><p>The discrimination seems better than with classical LDA based on all genes. This is very likely caused by too much noise in the full data matrix with over 20000 predictors.</p></li>
</ul>
<pre class="r"><code># BiocManager::install(&quot;sparseLDA&quot;)

library(sparseLDA)</code></pre>
<pre><code>## 
## Attaching package: &#39;sparseLDA&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     normalize</code></pre>
<pre class="r"><code>YDummy &lt;- data.frame(
  Y1 = ifelse(Y == 1, 1, 0),
  Y2 = ifelse(Y == 2, 1, 0),
  Y3 = ifelse(Y == 3, 1, 0)
  )

X2 &lt;- X[,1:150]

breast.slda &lt;- sda(x = X2,
  y = as.matrix(YDummy),
  lambda = 1e-6,
  stop = -50,
  maxIte = 25,
  trace = TRUE)</code></pre>
<pre><code>## ite:  1  ridge cost:  92.13987  |b|_1:  5.558217 
## ite:  2  ridge cost:  89.61364  |b|_1:  5.704252 
## ite:  3  ridge cost:  88.37327  |b|_1:  5.833096 
## ite:  4  ridge cost:  87.70409  |b|_1:  5.92812 
## ite:  5  ridge cost:  88.05903  |b|_1:  5.852153 
## ite:  6  ridge cost:  87.50389  |b|_1:  5.954857 
## ite:  7  ridge cost:  87.81367  |b|_1:  5.895391 
## ite:  8  ridge cost:  87.85366  |b|_1:  5.887668 
## ite:  9  ridge cost:  87.79671  |b|_1:  5.898385 
## ite:  10  ridge cost:  87.76715  |b|_1:  5.903966 
## ite:  11  ridge cost:  87.7727  |b|_1:  5.902912 
## ite:  12  ridge cost:  87.77724  |b|_1:  5.902051 
## ite:  13  ridge cost:  87.77948  |b|_1:  5.901627 
## ite:  14  ridge cost:  87.78059  |b|_1:  5.901418 
## ite:  15  ridge cost:  87.78113  |b|_1:  5.901315 
## ite:  16  ridge cost:  87.7814  |b|_1:  5.901265 
## ite:  17  ridge cost:  87.78153  |b|_1:  5.90124 
## ite:  18  ridge cost:  87.7816  |b|_1:  5.901227 
## ite:  1  ridge cost:  134.7777  |b|_1:  5.379261 
## ite:  2  ridge cost:  134.7777  |b|_1:  5.379261 
## final update, total ridge cost:  222.5593  |b|_1:  11.28049</code></pre>
<pre class="r"><code>Vsda &lt;- matrix(0, nrow=ncol(X2), ncol=2)
Vsda[breast.slda$varIndex,] &lt;- breast.slda$beta
colnames(Vsda) &lt;- paste0(&quot;V&quot;,1:ncol(Vsda))

Zsda &lt;- X2%*%Vsda
colnames(Zsda) &lt;- paste0(&quot;Z&quot;,1:ncol(Zsda))

Zsda %&gt;%
 as.data.frame %&gt;%
 mutate(grade = Y %&gt;% as.factor) %&gt;%
 ggplot(aes(x= Z1, y = Z2, color = grade)) +
 geom_point(size = 3) +
 ggtitle(&quot;sparse LDA on 150 genes&quot;)</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code>grid.arrange(
   Vsda %&gt;%
     as.data.frame %&gt;%
     mutate(geneID = 1:nrow(Vsda)) %&gt;%
     ggplot(aes(x = geneID, y = V1)) +
     geom_point(pch=21) +
     geom_hline(yintercept = c(-2,0,2)*sd(Vsda[,1]), col = &quot;red&quot;) ,

   Vsda %&gt;%
     as.data.frame %&gt;%
     mutate(geneID = 1:nrow(Vsda)) %&gt;%
     ggplot(aes(x = geneID, y = V2)) +
     geom_point(pch=21) +
     geom_hline(yintercept = c(-2,0,2)*sd(Vsda[,2]), col = &quot;red&quot;),
   ncol = 2)</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
</div>
<div id="lda-based-on-150-random-genes" class="section level2">
<h2><span class="header-section-number">4.3</span> LDA based on 150 random genes</h2>
<pre class="r"><code>breast.lda150 &lt;- MASS::lda(x = X2, grouping = Y)

Vlda &lt;- breast.lda150$scaling
colnames(Vlda) &lt;- paste0(&quot;V&quot;,1:ncol(Vlda))

Zlda &lt;- X2%*%Vlda
colnames(Zlda) &lt;- paste0(&quot;Z&quot;,1:ncol(Zlda))

grid.arrange(
  Zlda %&gt;%
    as.data.frame %&gt;%
    mutate(grade = Y %&gt;% as.factor) %&gt;%
    ggplot(aes(x= Z1, y = Z2, color = grade)) +
    geom_point(size = 3) +
    coord_fixed(),
  ggplot() +
    geom_bar(aes(x = c(&quot;z1&quot;,&quot;z2&quot;), y = breast.lda150$svd), stat = &quot;identity&quot;) +
    xlab(&quot;Discriminant&quot;) +
    ylab(&quot;Eigen Values&quot;),
  layout_matrix = matrix(
    c(1,1,2),
    nrow=1)
  )</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code>  grid.arrange(
    Vlda %&gt;%
      as.data.frame %&gt;%
      mutate(geneID = 1:nrow(Vlda)) %&gt;%
      ggplot(aes(x = geneID, y = V1)) +
      geom_point(pch=21) +
      geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,1]), col = &quot;red&quot;),

    Vlda %&gt;%
      as.data.frame %&gt;%
      mutate(geneID = 1:nrow(Vlda)) %&gt;%
      ggplot(aes(x = geneID, y = V2)) +
      geom_point(pch=21) +
      geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,2]), col = &quot;red&quot;),
    ncol = 2)</code></pre>
<p><img src="lda_files/figure-html/unnamed-chunk-6-2.png" width="672" /></p>
</div>
<div id="wrapup" class="section level2">
<h2><span class="header-section-number">4.4</span> Wrapup</h2>
<p>LDA on all 22283 genes gave poorer result than on 150 genes. This is probably caused by numerical instability when working with large <span class="math inline">\(\mathbf{W}\)</span> and <span class="math inline">\(\mathbf{B}\)</span> matrices</p>
<ul>
<li><p>Sparse LDA gave slightly poorer result than LDA on the subset of 150 genes. This may be caused by overfitting of the LDA.</p></li>
<li><p>When (sparse) LDA is used to build a prediction model/classifier, then CV methods, or splitting of dataset into training and test datasets should be used to allow for an honest evaluation of the final prediction model.</p></li>
<li><p>The graphs in the first two Fisher discriminant dimensions shown on the previous slides should only be used for data exploration.</p></li>
<li><p>When the objective is to try to understand differences between groups in a high dimensional space, Fisher LDA is preferred over PCA.</p></li>
</ul>
</div>
</div>
<div id="acknowledgement" class="section level1 unnumbered">
<h1>Acknowledgement</h1>
<ul>
<li>Olivier Thas for sharing his materials of Analysis of High Dimensional Data 2019-2020, which I used as the starting point for this chapter.</li>
</ul>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1>Session info</h1>
<details>
<p><summary>Session info</summary></p>
<pre><code>## [1] &quot;2021-10-20 13:36:43 UTC&quot;</code></pre>
<pre><code>## ─ Session info ───────────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 4.1.1 (2021-08-10)
##  os       macOS Catalina 10.15.7      
##  system   x86_64, darwin17.0          
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  ctype    en_US.UTF-8                 
##  tz       UTC                         
##  date     2021-10-20                  
## 
## ─ Packages ───────────────────────────────────────────────────────────────────
##  ! package           * version    date       lib
##  P AIMS              * 1.24.0     2021-05-19 [?]
##  P AnnotationDbi       1.54.1     2021-06-08 [?]
##  P assertthat          0.2.1      2019-03-21 [?]
##  P backports           1.2.1      2020-12-09 [?]
##  P Biobase           * 2.52.0     2021-05-19 [?]
##  P BiocFileCache       2.0.0      2021-05-19 [?]
##  P BiocGenerics      * 0.38.0     2021-05-19 [?]
##  P biomaRt           * 2.48.3     2021-08-15 [?]
##  P Biostrings          2.60.2     2021-08-05 [?]
##  P bit                 4.0.4      2020-08-04 [?]
##  P bit64               4.0.5      2020-08-30 [?]
##  P bitops              1.0-7      2021-04-24 [?]
##  P blob                1.2.2      2021-07-23 [?]
##  P bootstrap           2019.6     2019-06-17 [?]
##  P breastCancerMAINZ * 1.30.0     2021-05-20 [?]
##  P broom               0.7.9      2021-07-27 [?]
##  P bslib               0.3.0      2021-09-02 [?]
##  P cachem              1.0.6      2021-08-19 [?]
##  P cellranger          1.1.0      2016-07-27 [?]
##  P class               7.3-19     2021-05-03 [?]
##  P cli                 3.0.1      2021-07-17 [?]
##  P cluster           * 2.1.2      2021-04-17 [?]
##  P codetools           0.2-18     2020-11-04 [?]
##  P colorspace          2.0-2      2021-06-24 [?]
##  P crayon              1.4.1      2021-02-08 [?]
##  P curl                4.3.2      2021-06-23 [?]
##  P DBI                 1.1.1      2021-01-15 [?]
##  P dbplyr              2.1.1      2021-04-06 [?]
##  P digest              0.6.28     2021-09-23 [?]
##  P dplyr             * 1.0.7      2021-06-18 [?]
##  P e1071             * 1.7-9      2021-09-16 [?]
##  P elasticnet          1.3        2020-05-15 [?]
##  P ellipsis            0.3.2      2021-04-29 [?]
##  P evaluate            0.14       2019-05-28 [?]
##  P fansi               0.5.0      2021-05-25 [?]
##  P farver              2.1.0      2021-02-28 [?]
##  P fastmap             1.1.0      2021-01-25 [?]
##  P filelock            1.0.2      2018-10-05 [?]
##  P forcats           * 0.5.1      2021-01-27 [?]
##  P fs                  1.5.0      2020-07-31 [?]
##  P future              1.22.1     2021-08-25 [?]
##  P future.apply        1.8.1      2021-08-10 [?]
##  P genefu            * 2.24.2     2021-05-23 [?]
##  P generics            0.1.0      2020-10-31 [?]
##  P GenomeInfoDb        1.28.4     2021-09-09 [?]
##  P GenomeInfoDbData    1.2.6      2021-10-20 [?]
##  P ggplot2           * 3.3.5      2021-06-25 [?]
##  P globals             0.14.0     2020-11-22 [?]
##  P glue                1.4.2      2020-08-27 [?]
##  P gridExtra         * 2.3        2017-09-09 [?]
##  P gtable              0.3.0      2019-03-25 [?]
##  P haven               2.4.3      2021-08-04 [?]
##  P highr               0.9        2021-04-16 [?]
##  P hms                 1.1.1      2021-09-26 [?]
##  P htmltools           0.5.2      2021-08-25 [?]
##  P httr                1.4.2      2020-07-20 [?]
##  P iC10              * 1.5        2019-02-08 [?]
##  P iC10TrainingData  * 1.3.1      2018-08-24 [?]
##  P impute            * 1.66.0     2021-05-19 [?]
##  P IRanges             2.26.0     2021-05-19 [?]
##  P jquerylib           0.1.4      2021-04-26 [?]
##  P jsonlite            1.7.2      2020-12-09 [?]
##  P KEGGREST            1.32.0     2021-05-19 [?]
##  P KernSmooth          2.23-20    2021-05-03 [?]
##  P knitr               1.33       2021-04-24 [?]
##  P labeling            0.4.2      2020-10-20 [?]
##  P lars                1.2        2013-04-24 [?]
##  P lattice             0.20-45    2021-09-22 [?]
##  P lava                1.6.10     2021-09-02 [?]
##  P lifecycle           1.0.1      2021-09-24 [?]
##  P limma               3.48.3     2021-08-10 [?]
##  P listenv             0.8.0      2019-12-05 [?]
##  P lubridate           1.7.10     2021-02-26 [?]
##  P magrittr            2.0.1      2020-11-17 [?]
##  P MASS                7.3-54     2021-05-03 [?]
##  P Matrix              1.3-4      2021-06-01 [?]
##  P mclust              5.4.7      2020-11-20 [?]
##  P mda                 0.5-2      2020-06-29 [?]
##  P memoise             2.0.0      2021-01-26 [?]
##  P modelr              0.1.8      2020-05-19 [?]
##  P munsell             0.5.0      2018-06-12 [?]
##  P pamr              * 1.56.1     2019-04-22 [?]
##  P parallelly          1.28.1     2021-09-09 [?]
##  P pillar              1.6.3      2021-09-26 [?]
##  P pkgconfig           2.0.3      2021-10-20 [?]
##  P png                 0.1-7      2013-12-03 [?]
##  P prettyunits         1.1.1      2020-01-24 [?]
##  P prodlim           * 2019.11.13 2019-11-17 [?]
##  P progress            1.2.2      2019-05-16 [?]
##  P proxy               0.4-26     2021-06-07 [?]
##  P purrr             * 0.3.4      2020-04-17 [?]
##  P R6                  2.5.1      2021-08-19 [?]
##  P rappdirs            0.3.3      2021-01-31 [?]
##  P Rcpp                1.0.7      2021-07-07 [?]
##  P RCurl               1.98-1.4   2021-08-17 [?]
##  P readr             * 2.0.1      2021-08-10 [?]
##  P readxl              1.3.1      2019-03-13 [?]
##  P renv                0.14.0     2021-07-21 [?]
##  P reprex              2.0.1      2021-08-05 [?]
##  P rlang               0.4.11     2021-04-30 [?]
##  P rmarkdown           2.10       2021-08-06 [?]
##  P rmeta               3.0        2018-03-20 [?]
##  P RSQLite             2.2.8      2021-08-21 [?]
##  P rstudioapi          0.13       2020-11-12 [?]
##  P rvest               1.0.1      2021-07-26 [?]
##  P S4Vectors           0.30.0     2021-05-19 [?]
##  P sass                0.4.0      2021-05-12 [?]
##  P scales              1.1.1      2020-05-11 [?]
##  P sessioninfo         1.1.1      2018-11-05 [?]
##  P sparseLDA         * 0.1-9      2016-09-22 [?]
##  P stringi             1.7.4      2021-08-25 [?]
##  P stringr           * 1.4.0      2019-02-10 [?]
##  P SuppDists           1.1-9.5    2020-01-18 [?]
##  P survcomp          * 1.42.0     2021-05-19 [?]
##  P survival          * 3.2-13     2021-08-24 [?]
##  P survivalROC         1.0.3      2013-01-13 [?]
##  P tibble            * 3.1.5      2021-09-30 [?]
##  P tidyr             * 1.1.4      2021-09-27 [?]
##  P tidyselect          1.1.1      2021-04-30 [?]
##  P tidyverse         * 1.3.1      2021-04-15 [?]
##  P tzdb                0.1.2      2021-07-20 [?]
##  P utf8                1.2.2      2021-07-24 [?]
##  P vctrs               0.3.8      2021-04-29 [?]
##  P withr               2.4.2      2021-04-18 [?]
##  P xfun                0.25       2021-08-06 [?]
##  P XML                 3.99-0.8   2021-09-17 [?]
##  P xml2                1.3.2      2020-04-23 [?]
##  P XVector             0.32.0     2021-05-19 [?]
##  P yaml                2.2.1      2020-02-01 [?]
##  P zlibbioc            1.38.0     2021-05-19 [?]
##  source                          
##  Bioconductor                    
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  Bioconductor                    
##  Bioconductor                    
##  Bioconductor                    
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  standard (@1.0-7)               
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  standard (@0.3.2)               
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Github (r-lib/pkgconfig@b81ae03)
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  standard (@1.1.1)               
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  standard (@0.3.8)               
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.1)                  
##  CRAN (R 4.1.0)                  
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
##  CRAN (R 4.1.0)                  
##  Bioconductor                    
## 
## [1] /Users/runner/work/HDDA21/HDDA21/renv/library/R-4.1/x86_64-apple-darwin17.0
## [2] /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpIWQsJk/renv-system-library
## 
##  P ── Loaded and on-disk path mismatch.</code></pre>
</details>
</div>

<div id="rmd-source-code">---
title: "Linear Discriminant Analysis (LDA)"
author: "Lieven Clement"
date: "statOmics, Ghent University (https://statomics.github.io)"
output:
  pdf_document:
    toc: true
    number_sections: true
    latex_engine: xelatex
---

```{r}
library(tidyverse)
library(gridExtra)
```

# Breast cancer example

- Schmidt *et al.*, 2008, Cancer Research, **68**, 5405-5413

- Gene expression patterns in n=200 breast tumors were investigated (p=22283 genes)

- After surgery the tumors were graded by a pathologist (stage 1,2,3)


## Data

```{r, message=FALSE, warning=FALSE}
#BiocManager::install("genefu")
#BiocManager::install("breastCancerMAINZ")

library(genefu)
library(breastCancerMAINZ)
data(mainz)

X <- t(exprs(mainz)) # gene expressions
n <- nrow(X)
H <- diag(n)-1/n*matrix(1,ncol=n,nrow=n)
X <- H%*%X
Y <- pData(mainz)$grade
table(Y)

svdX <- svd(X)
k <- 2
Zk <- svdX$u[,1:k] %*% diag(svdX$d[1:k])
colnames(Zk) <- paste0("Z",1:k)

Zk %>%
  as.data.frame %>%
  mutate(grade = Y %>% as.factor) %>%
  ggplot(aes(x= Z1, y = Z2, color = grade)) +
  geom_point(size = 3)
```

# Linear discriminant analysis

Fisher's construction of LDA is simple: it allows for classification in a dimension-reduced subspace of $\mathbb{R}^p$.

  First we assume that $Y$ can only take two values (0/1).

  Fisher aimed for a direction, say $\mathbf{a}$, in the $p$-dimensional predictor space such that the orthogonal projections of the predictors, $\mathbf{x}^t\mathbf{a}$, show maximal ratio between the between and within sums of squares:
  \[
  \mathbf{v} = \text{ArgMax}_a \frac{\mathbf{a}^t\mathbf{B}\mathbf{a}}{\mathbf{a}^t\mathbf{W}\mathbf{a}} \text{ subject to } {\mathbf{a}^t\mathbf{W}\mathbf{a}}=1,
 \]
 where $\mathbf{W}$ and $\mathbf{B}$ are the within and between covariance matrices of $\mathbf{x}$.
 The restriction is introduced to obtain a (convenient) unique solution.

---

## Between and within sums of squares


- In the training dataset, let $\mathbf{x}_{ik}$ denote the $i$th $p$-dimensional observation in the $k$th group ($k=0,1$ referring to $Y=0$ and $Y=1$, resp.), $i=1,\ldots, n_k$.

- Let $z_{ik}=\mathbf{a}^t\mathbf{x}_{ik}$ denote the orthogonal projection of $\mathbf{x}_{ik}$ onto $\mathbf{a}$

- For the one-dimensional $z$-observations, consider the following sum of squares:
  \[
    \text{SSE}=\text{within sum of squares} = \sum_{k=0,1}\sum_{i=1}^{n_k} (z_{ik}-\bar{z}_k)^2
  \]
  \[
    \text{SSB}=\text{between sum of squares} = \sum_{k=0,1}\sum_{i=1}^{n_k} (\bar{z}_{k}-\bar{z})^2 = \sum_{k=0,1} n_k (\bar{z}_{k}-\bar{z})^2
  \]
  with $\bar{z}_k$ the sample mean of $z_{ik}$ within group $k$, and $\bar{z}$ the sample mean of all $z_{ik}$.

- To reformulate SSE and SSB in terms of the $p$-dimensional $\mathbf{x}_{ik}$, we need the sample means
  \[
    \bar{z}_k = \frac{1}{n_k} \sum_{i=1}^{n_k} z_{ik} = \frac{1}{n_k} \sum_{i=1}^{n_k} \mathbf{a}^t\mathbf{x}_{ik} = \mathbf{a}^t \frac{1}{n_k} \sum_{i=1}^{n_k} \mathbf{x}_{ik} = \mathbf{a}^t \bar{\mathbf{x}}_k
  \]
  \[
    \bar{z} = \frac{1}{n}\sum_{k=0,1}\sum_{i=1}^{n_k} z_{ik} = \cdots = \mathbf{a}^t\bar{\mathbf{x}}.
  \]

- SSE becomes
  \[
    \text{SSE} = \sum_{k=0,1}\sum_{i=1}^{n_k} (z_{ik}-\bar{z}_k)^2 = \mathbf{a}^t \left(\sum_{k=0,1}\sum_{i=1}^{n_k} (\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)(\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)^t\right)\mathbf{a}
  \]

- SSB becomes
  \[
    \text{SSB} = \sum_{k=0,1} n_k (\bar{z}_{k}-\bar{z})^2 = \mathbf{a}^t \left(\sum_{k=0,1} n_k (\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})(\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})^t \right)\mathbf{a}
  \]

---

- The $p \times p$ matrix
   \[
    \mathbf{W}=\sum_{k=0,1}\sum_{i=1}^{n_k} (\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)(\mathbf{x}_{ik}-\bar{\mathbf{x}}_k)^t
   \]
   is referred to as the matrix of **within** sum of squares and cross products.

- The $p \times p$ matrix
   \[
    \mathbf{B}=\sum_{k=0,1} n_k (\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})(\bar{\mathbf{x}_{k}}-\bar{\mathbf{x}})^t
   \]
   is referred to as the matrix of **between** sum of squares and cross products.

- Note that on the diagonal of $\mathbf{W}$ and $\mathbf{B}$ you find the ordinary univariate within and between sums of squares of the individual components of $\mathbf{x}$.

## Obtain projections

An equivalent formulation:
  \[
  \mathbf{v} = \text{ArgMax}_a \mathbf{a}^t\mathbf{B}\mathbf{a} \text{ subject to } \mathbf{a}^t\mathbf{W}\mathbf{a}=1.
 \]

 This can be solved by introducing a Langrange multiplier:
  \[
  \mathbf{v} = \text{ArgMax}_a \mathbf{a}^t\mathbf{B}\mathbf{a} -\lambda(\mathbf{a}^t\mathbf{W}\mathbf{a}-1).
 \]

 Calculating the partial derivative w.r.t. $\mathbf{a}$ and setting it to zero gives
 \begin{eqnarray*}
   2\mathbf{B}\mathbf{a} -2\lambda \mathbf{W}\mathbf{a} &=& 0\\
   \mathbf{B}\mathbf{a}  &=& \lambda \mathbf{W}\mathbf{a} \\
   \mathbf{W}^{-1}\mathbf{B}\mathbf{a} &=& \lambda\mathbf{a}.
 \end{eqnarray*}

 From the final equation we recognise that $\mathbf{v}=\mathbf{a}$ is an **eigenvector** of $\mathbf{W}^{-1}\mathbf{B}$, and $\lambda$ is the corresponding **eigenvalue**.

 The equation has in general $\text{rank}(\mathbf{W}^{-1}\mathbf{B})$ solutions. In the case of two classes, the rank equals 1 and thus only one solution exists.

---

- A training data set is used for the calculation of $\mathbf{W}$ and $\mathbf{B}$. $\longrightarrow$ This gives the eigenvector $\mathbf{v}$

- The training data is also used for the calculation of the centroids of the classes (e.g. the sample means, say $\bar{\mathbf{x}}_1$ and $\bar{\mathbf{x}}_2$). $\longrightarrow$ The projected centroids are given by $\bar{\mathbf{x}}_1^t\mathbf{v}$ and $\bar{\mathbf{x}}_2^t\mathbf{v}$.

- A new observation with predictor $\mathbf{x}$ is classified in the class for which the projected centroid is closest to the projected predictor $z=\mathbf{x}^t\mathbf{v}$.

An advantage of this approach is that $\mathbf{v}$ can be interpreted (similar as the loadings in a PCA) in terms of which predictors $x_j$ are important to discriminate between classes 0 and 1.

---

## More than two classes

When the outcome $Y$ refers to more than two classes, say $m$ classes, then Fisher's method is constructed in exactly the same way. Now
\[
  \mathbf{W}^{-1}\mathbf{B}\mathbf{a} = \lambda\mathbf{a}
\]
will have $r=\text{rank}(\mathbf{W}^{-1}\mathbf{B}) = \min(m-1,p,n)$ solutions (eigenvectors and eigenvalues). ($n$: sample size of training data)

Let $\mathbf{v}_j$ and $\lambda_j$ denote the $r$ solutions, and define

- $\mathbf{V}$: $p\times r$ matrix with collums $\mathbf{v}$

- $\mathbf{L}$: $r \times r$ diagonal matrix with elements $\lambda_1 > \lambda_2 > \cdots > \lambda_r$

The $p$-dimensional predictor data in $\mathbf{X}$ may then be transformed to the $r$-dimensional scores
\[
  \mathbf{Z} = \mathbf{X}\mathbf{V}.
\]

---

For eigenvectors $\mathbf{v}_i$ and $\mathbf{v}_j$, it holds that
\[
 \text{cov}\left[Z_i,Z_j\right] = \text{cov}\left[\mathbf{X}\mathbf{v}_i,\mathbf{X}\mathbf{v}_j\right]= \mathbf{v}_i^t \mathbf{W} \mathbf{v}_j = \delta_{ij} ,
\]
in which the covariances are defined within groups. Hence, within the groups (classes) the scores are uncorrelated.

# High dimensional predictors

With high-dimensional predictors

- Replace the $p\times p$ matrices $\mathbf{W}$ and $\mathbf{B}$ by their diagonal matrices (i.e. put zeroes on the off-diagonal positions)

-  **Sparse LDA** by imposing an $L_1$-penalty on $\mathbf{v}$.

Two approaches: Zhou *et al.* (2006), Journal of Computational and Graphical Statistics , **15**, 265-286, and [Clemmensen *et al.* (2011), Technometrics, **53**](https://web.stanford.edu/~hastie/Papers/sda_resubm_daniela-final.pdf).

# Breast cancer example

## All genes

### LDA

- Fisher's LDA is illustrated on the breast cancer data with all three tumor stages as outcome.

- We try to discriminate between the different stages according to the gene expression data of all genes.

- We cache the result because the calculation takes 10 minutes.

```{r run-lda, cache=TRUE}
breast.lda <- MASS::lda(x = X, grouping = Y)
```

```{r}
Vlda <- breast.lda$scaling
colnames(Vlda) <- paste0("V",1:ncol(Vlda))

Zlda <- X%*%Vlda
colnames(Zlda) <- paste0("Z",1:ncol(Zlda))

grid.arrange(
  Zlda %>%
    as.data.frame %>%
    mutate(grade = Y %>% as.factor) %>%
    ggplot(aes(x= Z1, y = Z2, color = grade)) +
    geom_point(size = 3) +
    coord_fixed(),
  ggplot() +
    geom_bar(aes(x = c("z1","z2"), y = breast.lda$svd), stat = "identity") +
    xlab("Discriminant") +
    ylab("Eigen Values"),
  layout_matrix = matrix(
    c(1,1,2),
    nrow=1)
  )
```

---

- The columns of the matrix $\mathbf{V}$ contain the eigenvectors. There are $\min(3-2,22283,200)=2$ eigenvectors. The $200\times 2$ matrix $\mathbf{Z}$ contains the scores on the two Fisher discriminants.

- The eigenvalue $\lambda_j$ can be interpreted as the ratio
\[
  \frac{\mathbf{v}_j^t\mathbf{B}\mathbf{v}_j}{\mathbf{v}_j^t\mathbf{W}\mathbf{v}_j} ,
\]
or (upon using $\mathbf{v}_j^t\mathbf{W}\mathbf{v}_j=1$) the between-centroid sum of squares (in the reduced dimension space of the Fisher discriminants)
\[
  \mathbf{v}_j^t\mathbf{B}\mathbf{v}_j.
\]

- From the screeplot of the eigenvalues we see that the first dimension is more important than the second (not hugely) in terms of discriminating between the groups.

- From the scatterplot we can see that there is no perfect separation (discrimination) between the three tumor stages (quite some overlap).

- To some extent the first Fisher discriminant dimension discriminates stage 3 (green dots) from the other two stages, and the second dimension separates stage 1 (black dots) from the two others.

### Interpretation of loadings

```{r}
grid.arrange(
  Vlda %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vlda)) %>%
    ggplot(aes(x = geneID, y = V1)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,1]), col = "red"),

  Vlda %>%
    as.data.frame %>%
    mutate(geneID = 1:nrow(Vlda)) %>%
    ggplot(aes(x = geneID, y = V2)) +
    geom_point(pch=21) +
    geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,2]), col = "red"),
  ncol = 2)
```

The loadings of the Fisher discriminants are within the columns of the $\mathbf{V}$ matrix.

- Since we have 22283 genes, each discriminant is a linear combination of 22283 gene expression. Instead of looking at the listing of 22283 loadings, we made an index plot (no particular ordering of genes on horizontal axis).

- The red horizontal reference lines correspond to the average of the loading (close to zero) and the average plus and minus twice the standard deviation of the loadings.

- If no genes had any "significant" discriminating power, then we would expect approximately $95\%$ of all loadings within the band. Thus loadings outside of the band are of potential interest and may perhaps be discriminating between the three tumor stages.

- In the graphs presented here we see many loadings within the bands, but also many outside of the band.

We repeat the analysis, but now with the sparse LDA method of Clemmensen et al. (2011).

---

## Sparse LDA based on 150 random genes

- We only present the results of the sparse LDA based on a random subset of 150 genes (ordering of genes in datamatrix is random).

- The discrimination seems better than with classical LDA based on all genes. This is very likely caused by too much noise in the full data matrix with over 20000 predictors.

```{r}
# BiocManager::install("sparseLDA")

library(sparseLDA)

YDummy <- data.frame(
  Y1 = ifelse(Y == 1, 1, 0),
  Y2 = ifelse(Y == 2, 1, 0),
  Y3 = ifelse(Y == 3, 1, 0)
  )

X2 <- X[,1:150]

breast.slda <- sda(x = X2,
  y = as.matrix(YDummy),
  lambda = 1e-6,
  stop = -50,
  maxIte = 25,
  trace = TRUE)

Vsda <- matrix(0, nrow=ncol(X2), ncol=2)
Vsda[breast.slda$varIndex,] <- breast.slda$beta
colnames(Vsda) <- paste0("V",1:ncol(Vsda))

Zsda <- X2%*%Vsda
colnames(Zsda) <- paste0("Z",1:ncol(Zsda))

Zsda %>%
 as.data.frame %>%
 mutate(grade = Y %>% as.factor) %>%
 ggplot(aes(x= Z1, y = Z2, color = grade)) +
 geom_point(size = 3) +
 ggtitle("sparse LDA on 150 genes")

grid.arrange(
   Vsda %>%
     as.data.frame %>%
     mutate(geneID = 1:nrow(Vsda)) %>%
     ggplot(aes(x = geneID, y = V1)) +
     geom_point(pch=21) +
     geom_hline(yintercept = c(-2,0,2)*sd(Vsda[,1]), col = "red") ,

   Vsda %>%
     as.data.frame %>%
     mutate(geneID = 1:nrow(Vsda)) %>%
     ggplot(aes(x = geneID, y = V2)) +
     geom_point(pch=21) +
     geom_hline(yintercept = c(-2,0,2)*sd(Vsda[,2]), col = "red"),
   ncol = 2)
```

## LDA based on 150 random genes

```{r}
breast.lda150 <- MASS::lda(x = X2, grouping = Y)

Vlda <- breast.lda150$scaling
colnames(Vlda) <- paste0("V",1:ncol(Vlda))

Zlda <- X2%*%Vlda
colnames(Zlda) <- paste0("Z",1:ncol(Zlda))

grid.arrange(
  Zlda %>%
    as.data.frame %>%
    mutate(grade = Y %>% as.factor) %>%
    ggplot(aes(x= Z1, y = Z2, color = grade)) +
    geom_point(size = 3) +
    coord_fixed(),
  ggplot() +
    geom_bar(aes(x = c("z1","z2"), y = breast.lda150$svd), stat = "identity") +
    xlab("Discriminant") +
    ylab("Eigen Values"),
  layout_matrix = matrix(
    c(1,1,2),
    nrow=1)
  )

  grid.arrange(
    Vlda %>%
      as.data.frame %>%
      mutate(geneID = 1:nrow(Vlda)) %>%
      ggplot(aes(x = geneID, y = V1)) +
      geom_point(pch=21) +
      geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,1]), col = "red"),

    Vlda %>%
      as.data.frame %>%
      mutate(geneID = 1:nrow(Vlda)) %>%
      ggplot(aes(x = geneID, y = V2)) +
      geom_point(pch=21) +
      geom_hline(yintercept = c(-2,0,2)*sd(Vlda[,2]), col = "red"),
    ncol = 2)
  ```

## Wrapup

LDA on all 22283 genes gave poorer result than on 150 genes.
This is probably caused by numerical instability when working with large $\mathbf{W}$ and $\mathbf{B}$ matrices

- Sparse LDA gave slightly poorer result than LDA on the subset of 150 genes. This may be caused by overfitting of the LDA.

- When (sparse) LDA is used to build a prediction model/classifier, then CV methods, or splitting of dataset into training and test datasets should be used to allow for an honest evaluation of the final prediction model.

- The graphs in the first two Fisher discriminant dimensions shown on the previous slides should only be used for data exploration.

- When the objective is to try to understand differences between groups in a high dimensional space, Fisher LDA is preferred over PCA.


# Acknowledgement {-}

- Olivier Thas for sharing his materials of Analysis of High Dimensional Data 2019-2020, which I used as the starting point for this chapter.

# Session info {-}

<details><summary>Session info</summary>

```{r session_info, echo=FALSE, cache=FALSE}
Sys.time()
sessioninfo::session_info()
```

</details>
</div>
<div class="footer">
    <hr>
    This work is licensed under the <a href= "https://creativecommons.org/licenses/by-nc-sa/4.0">
    CC BY-NC-SA 4.0</a> licence.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("lda.Rmd");
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
